{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Zoonotic Potential"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [*] remember to turn of EC2 instance\n",
    "\n",
    "\n",
    "### Plan\n",
    "\n",
    "- measure host abundance for each zoonotic SOTU\n",
    "- measure host abundance for each non-zoonotic SOTU\n",
    "- measure differential between positive and negative sets\n",
    "- statistical test of differential AGS with FGS of zoonotic potential (Virus set enrichment analysis)\n",
    "\n",
    "### Done\n",
    "\n",
    "- [x] parse virus dataset\n",
    "- [x] find number of genuses with differing in zoonotic values\n",
    "- [x] find rank distribution of has_potential_host\n",
    "    - total: 438167, species: 414190, genus: 24, family: 49\n",
    "- [x] manually patch tax_ids for all species\n",
    "- [x] find number of virus species with associated SOTU\n",
    "    - match genus taxons via list\n",
    "- [x] create projection of SOTUs and sequence alignments using zoonotic viruses\n",
    "- [x] setup graphistry visualization\n",
    "- [x] clustering analysis on SOTUs\n",
    "    - [x] setup community detection with seeded taxId\n",
    "    - [x] WCC\n",
    "    - [x] Louvain and lieden\n",
    "- [x] Include STAT hosts into projection \n",
    "\n",
    "### TODO\n",
    "\n",
    "- [*] understand discrepancy between zoonotic label and human host association\n",
    "    - how many non-zoonotic viruses have human stat associations\n",
    "    - goal: may need to re-examine using dataset for FGS?\n",
    "    - may be able to create our own zoonotic dataset using STAT?\n",
    "    - create a functional gene set (FGS) using entirety of the network of viruses in humans? if there is strong agreement\n",
    "    - tests:\n",
    "        - check edges for any viruses with human stat host > threshold, check IsZoonotic values\n",
    "        - cluster using human vs other\n",
    "- [*] update Graphistry with stat hosts (order level only)\n",
    "- [*] test vary seq alignment threshold > 90 in clustering, \n",
    "    - goal: may not need to cluster SOTUs into viral communities\n",
    "- [*] examine weight/threshold on HAS_POTENTIAL_TAXON > 80\n",
    "    - goal: may not want to overly rely on tax label and use similar clusters \n",
    "- Decide on including tissues\n",
    "    - check  \n",
    "- decide on using community or individual virus order/species\n",
    "    - find differentially expression of hosts (AGS) between positive and negative sets\n",
    "    - compare with FGS from entire network with specific hosts?\n",
    "- for each {community, species} find abundance of host taxons\n",
    "    - remap host abundance by class [mammals, fish, plants, fungii, parasite]\n",
    "    - find taxids for each\n",
    "- for each {community, species} measure network topology metric of sotus and full trait hierachy\n",
    "    - Modularity, etc\n",
    "    - spectrum\n",
    "- compare variance within zoonotic and non-zoonotic\n",
    "    - log intensity-ratio\n",
    "- compare differential change between zoonotic and non-zoonotic\n",
    "    - rank-sum test \n",
    "    - https://nbisweden.github.io/excelerate-scRNAseq/session-de/session-de-methods.html\n",
    "- for each community, find tissue abundance by class\n",
    "\n",
    "- create projection of SOTUs and sequence alignments using non-zoonotic viruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lukepereira/anaconda3/envs/rnalab/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "# Notebook config\n",
    "import sys\n",
    "if '../' not in sys.path:\n",
    "    sys.path.append(\"../\")\n",
    "%load_ext dotenv\n",
    "%reload_ext dotenv\n",
    "%dotenv\n",
    "\n",
    "import collections\n",
    "import os\n",
    "import urllib.parse\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import graphistry\n",
    "\n",
    "\n",
    "from datasources.neo4j import gds\n",
    "from queries import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<graphdatascience.graph_data_science.GraphDataScience object at 0x11d753cd0>\n"
     ]
    }
   ],
   "source": [
    "base_data_path = './zoonosis_data/'\n",
    "virus_dataset = base_data_path + 'trefle/'\n",
    "neo4j_data_path = base_data_path + 'neo4j/'\n",
    "graphistry_data_path = base_data_path + 'graphistry/'\n",
    "\n",
    "\n",
    "graphistry.register(\n",
    "    api=3,\n",
    "    username=os.getenv('GRAPHISTRY_USERNAME'),\n",
    "    password=os.getenv('GRAPHISTRY_PASSWORD'),\n",
    ")\n",
    "\n",
    "print(gds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(virus_dataset + 'viruses.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fill in tax_ids (only run once)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_names = df['vVirusNameCorrected'].unique()\n",
    "\n",
    "def parse_scientific_name(name):\n",
    "    \"\"\"\n",
    "    Parse a scientific name into its components.\n",
    "    \"\"\"\n",
    "    name = name.replace('_', ' ')\n",
    "    return name\n",
    "\n",
    "with open( virus_dataset + 'species_names.txt', 'w') as f:\n",
    "    for line in unique_names:\n",
    "        f.write(f\"{parse_scientific_name(line)}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "cat ./zoonosis_data/trefle/species_names.txt | taxonkit name2taxid  >> ./zoonosis_data/trefle/species_taxids.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                name    taxid\n",
      "0               Adelaide River virus  31612.0\n",
      "1           Adeno-associated virus-1      NaN\n",
      "2           Adeno-associated virus-2      NaN\n",
      "3           Adeno-associated virus-5      NaN\n",
      "4  African green monkey polyomavirus  12480.0\n",
      "(586, 2)\n",
      "269\n"
     ]
    }
   ],
   "source": [
    "df2 = pd.read_csv(virus_dataset + 'species_taxids.tsv', sep='\\t', header=None, names=['name', 'taxid'])\n",
    "print(df2.head())\n",
    "print(df2.shape)\n",
    "print(df2['taxid'].isna().sum())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load zoonotic dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22,)\n",
      "(586, 24)\n",
      "(564, 24)\n",
      "\n",
      "0    378\n",
      "1    186\n",
      "Name: IsZoonotic, dtype: int64\n",
      "0    458\n",
      "1    106\n",
      "Name: IsZoonotic.stringent, dtype: int64\n",
      "0    557\n",
      "1      7\n",
      "Name: ReverseZoonoses, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/vl/wsxpm_412lgcfm03dqd31qmh0000gn/T/ipykernel_3501/3246926767.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(int)\n",
      "/var/folders/vl/wsxpm_412lgcfm03dqd31qmh0000gn/T/ipykernel_3501/3246926767.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(str)\n"
     ]
    }
   ],
   "source": [
    "# Manually fill in any unresolved taxids\n",
    "df2 = pd.read_csv(virus_dataset + 'species_taxids_patched.tsv', sep='\\t', header=None, names=['name', 'taxid'])\n",
    "df2['name'] = df2['name'].str.replace(' ', '_')\n",
    "missing_taxid = df2[df2['taxid'].isna()]['name']\n",
    "\n",
    "print(missing_taxid.shape)\n",
    "\n",
    "merged = pd.merge(df, df2, left_on='vVirusNameCorrected', right_on='name', how='left')\n",
    "merged = merged.drop(columns=['name'])\n",
    "merged = merged.rename(columns={'taxid': 'TaxID'})\n",
    "merged_dropna = merged.dropna(subset=['TaxID'])\n",
    "merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(int)\n",
    "merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(str)\n",
    "\n",
    "print(merged.shape)\n",
    "print(merged_dropna.shape)\n",
    "print()\n",
    "\n",
    "print(merged_dropna['IsZoonotic'].value_counts())\n",
    "print(merged_dropna['IsZoonotic.stringent'].value_counts())\n",
    "print(merged_dropna['ReverseZoonoses'].value_counts())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Neo4j queries"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dataset Serratus overlap QA:\n",
    "- 564 virus in original dataset, 513 after parsing name to ncbi taxid (mapped node in graph)\n",
    "- 181 virus taxons have associated SOTU, 383 do not\n",
    "- 104 are not zoonotic, 77 are zoonotic\n",
    "- Including all palmprints instead of only SOTUs produces the same edges except for virus taxon 92129 (Tupaia paramyxovirus) which has a large number of palmprints along with SOTU u15499. To keep things simple, I only include associations between SOTUs and virus taxons\n",
    "- Including children and parents of virus taxons only increases SOTU matches by 1 to 182 SOTUs. To keep things simple, I only include direct matches to species in the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_sotu_nodes = \"\"\"\n",
    "    MATCH (a:SOTU)-[:HAS_POTENTIAL_TAXON]->(b:Taxon)\n",
    "    WHERE b.taxId in $tax_ids\n",
    "    RETURN\n",
    "        id(a) as nodeId,\n",
    "        a.palmId as appId,\n",
    "        a.palmId as palmId,\n",
    "        labels(a) as labels,\n",
    "        b.taxId as taxId,\n",
    "        b.rank as taxRank,\n",
    "        b.taxOrder as taxOrder,\n",
    "        CASE WHEN b.taxId in $zoonotic_tax_ids THEN True ELSE False END AS isZoonotic\n",
    "\"\"\"\n",
    "\n",
    "query_sotu_msa_edges = \"\"\"\n",
    "    MATCH (a:SOTU)-[r:SEQUENCE_ALIGNMENT]->(b:SOTU)\n",
    "    WHERE a.palmId in $sotus\n",
    "    AND b.palmId in $sotus\n",
    "    RETURN\n",
    "        id(a) as sourceNodeId,\n",
    "        a.palmId as sourceAppId,\n",
    "        id(b) as targetNodeId,\n",
    "        b.palmId as targetAppId,\n",
    "        'SEQUENCE_ALIGNMENT' as relationshipType,\n",
    "        r.percentIdentity as percentIdentity,\n",
    "        r.percentIdentity as weight\n",
    "\"\"\"\n",
    "\n",
    "query_stat_taxon_order_nodes = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "    }\n",
    "    WITH t, u\n",
    "    RETURN\n",
    "        id(t) as nodeId,\n",
    "        t.taxId as appId,\n",
    "        t.taxId as taxId,\n",
    "        labels(t) as labels,\n",
    "        t.rank as taxRank,\n",
    "        t.taxOrder as taxOrder,\n",
    "        CASE WHEN u.taxId = '40674' THEN 'Mammal' ELSE \n",
    "            CASE WHEN (u.taxId = '1476529' OR u.taxId = '7777' OR u.taxID = '1476750') THEN 'Fish' ELSE \n",
    "                CASE WHEN u.taxId = '33090' THEN 'Plant' ELSE \n",
    "                    CASE WHEN u.taxId = '4751' THEN 'Fungi' ELSE \n",
    "                        CASE When u.taxId = '33630' THEN 'Parasite' ELSE \n",
    "                            CASE WHEN u.taxId = '2' THEN 'Bacteria' ELSE \n",
    "                                CASE WHEN u.taxId = '10239' THEN 'Virus' ELSE 'Other'\n",
    "                                END\n",
    "                            END\n",
    "                        END\n",
    "                    END\n",
    "                END\n",
    "            END\n",
    "        END AS hostClass\n",
    "\"\"\"\n",
    "\n",
    "query_has_host_order_stat = '''\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, s, q\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, s, q\n",
    "    }\n",
    "    WITH p, t, r, s, q, avg(q.percentIdentityFull) as percentIdentityFull\n",
    "    RETURN\n",
    "        id(p) as sourceNodeId,\n",
    "        p.palmId as sourceAppId,\n",
    "        CASE WHEN percentIdentityFull >= 0.2 THEN id(t) ELSE 8765758 END AS targetNodeId,\n",
    "        CASE WHEN percentIdentityFull >= 0.2 THEN t.taxId ELSE '12908' END AS targetAppId,\n",
    "        'HAS_HOST_STAT' as relationshipType,\n",
    "        count(*) AS directAssociations,\n",
    "        count(*) AS count,\n",
    "        avg(r.percentIdentity) as avgPercentIdentityPalmprint,\n",
    "        avg(q.percentIdentity) as avgPercentIdentityStatKmers,\n",
    "        avg(q.percentIdentityFull) as avgPercentIdentityStatSpots,\n",
    "        avg(q.percentIdentity) * avg(r.percentIdentity) as weight\n",
    "'''\n",
    "\n",
    "### QA queries\n",
    "\n",
    "\n",
    "query_stat_taxon_nodes = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "    }\n",
    "    WITH t, u\n",
    "    RETURN\n",
    "        id(t) as nodeId,\n",
    "        t.taxId as appId,\n",
    "        t.taxId as taxId,\n",
    "        labels(t) as labels,\n",
    "        t.rank as taxRank,\n",
    "        t.taxOrder as taxOrder,\n",
    "        CASE WHEN u.taxId = '40674' THEN 'Mammal' ELSE \n",
    "            CASE WHEN (u.taxId = '1476529' OR u.taxId = '7777' OR u.taxID = '1476750') THEN 'Fish' ELSE \n",
    "                CASE WHEN u.taxId = '33090' THEN 'Plant' ELSE \n",
    "                    CASE WHEN u.taxId = '4751' THEN 'Fungi' ELSE \n",
    "                        CASE When u.taxId = '33630' THEN 'Parasite' ELSE \n",
    "                            CASE WHEN u.taxId = '2' THEN 'Bacteria' ELSE \n",
    "                                CASE WHEN u.taxId = '10239' THEN 'Virus' ELSE 'Other'\n",
    "                                END\n",
    "                            END\n",
    "                        END\n",
    "                    END\n",
    "                END\n",
    "            END\n",
    "        END AS hostClass\n",
    "\"\"\"\n",
    "\n",
    "query_has_host_stat = '''\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, s, q\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, s, q\n",
    "    }\n",
    "    WITH p, t, r, s, q, avg(q.percentIdentityFull) as percentIdentityFull\n",
    "    RETURN\n",
    "        id(p) as sourceNodeId,\n",
    "        p.palmId as sourceAppId,\n",
    "        CASE WHEN percentIdentityFull >= 0.2 THEN id(t) ELSE 8765758 END AS targetNodeId,\n",
    "        CASE WHEN percentIdentityFull >= 0.2 THEN t.taxId ELSE '12908' END AS targetAppId,\n",
    "        'HAS_HOST_STAT' as relationshipType,\n",
    "        count(*) AS directAssociations,\n",
    "        count(*) AS count,\n",
    "        avg(r.percentIdentity) as avgPercentIdentityPalmprint,\n",
    "        avg(q.percentIdentity) as avgPercentIdentityStatKmers,\n",
    "        avg(q.percentIdentityFull) as avgPercentIdentityStatSpots,\n",
    "        avg(q.percentIdentity) * avg(r.percentIdentity) as weight\n",
    "'''\n",
    "\n",
    "query_tax_ids_exist = \"\"\"\n",
    "    MATCH (a:Taxon)\n",
    "    WHERE a.taxId in $tax_ids\n",
    "    RETURN COLLECT(DISTINCT a.taxId) as tax_ids\n",
    "\"\"\"\n",
    "\n",
    "query_palmprint_exists = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (a:SOTU)-[:HAS_POTENTIAL_TAXON]->(b:Taxon)\n",
    "        WHERE b.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as palmprints\n",
    "        UNION\n",
    "        MATCH (a:SOTU)<-[:HAS_SOTU*]-(b:Palmprint)-[:HAS_POTENTIAL_TAXON]->(c:Taxon)\n",
    "        WHERE c.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) + COLLECT(DISTINCT b.palmId) as palmprints\n",
    "    }\n",
    "    WITH tax_id, palmprints\n",
    "    RETURN\n",
    "        tax_id,\n",
    "        palmprints\n",
    "\"\"\"\n",
    "\n",
    "query_sotus_exist_parent_child = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (a:SOTU)-[:HAS_POTENTIAL_TAXON]->(b:Taxon)\n",
    "        WHERE b.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as sotus\n",
    "        UNION\n",
    "        MATCH (a:Taxon)<-[:HAS_PARENT*]-(b:Taxon)<-[:HAS_POTENTIAL_TAXON]-(c:SOTU)\n",
    "        WHERE a.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as sotus\n",
    "        UNION\n",
    "        MATCH (a:Taxon)-[:HAS_PARENT*]->(b:Taxon)<-[:HAS_POTENTIAL_TAXON]-(c:SOTU)\n",
    "        WHERE a.taxId in $tax_ids\n",
    "        AND b.rank = 'species'\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as sotus\n",
    "    }\n",
    "    WITH tax_id, sotus\n",
    "    RETURN\n",
    "        tax_id,\n",
    "        sotus\n",
    "\"\"\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Neo4j dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_ids = list(merged_dropna.TaxID.unique())\n",
    "zoonotic_tax_ids = list(merged_dropna[merged_dropna['IsZoonotic'] == True].TaxID.unique())\n",
    "host_class_tax_ids = [\n",
    "    '40674', # [Mammal] Mammalia\n",
    "    '1476529', # [Fish], Cyclostomata \n",
    "    '7777' # [Fish], Chondrichthyes\n",
    "    '1476750', # [Fish], fish environmental sample\n",
    "    '33090', # [Plant], Viridiplantae\n",
    "    '4751', # [Fungi]\n",
    "    '33630', # [Parasite], Alveolata\n",
    "    '2', # [Bacteria]\n",
    "    '10239', # [Virus], Viruses\n",
    "]\n",
    "\n",
    "def _log_df(df):\n",
    "    namespace = globals()\n",
    "    var_name = [name for name in namespace if namespace[name] is df]\n",
    "    print(var_name, df.shape)\n",
    "    print(df.head())\n",
    "\n",
    "\n",
    "def fetch_cached_df(query, params, filename, use_cache=True, log=False):\n",
    "    if os.path.exists(neo4j_data_path + filename) and use_cache:\n",
    "        df = pd.read_csv(neo4j_data_path + filename)\n",
    "        df = utils.deserialize_df(df)\n",
    "    else:\n",
    "        df = gds.run_cypher(query, params=params)\n",
    "        df.to_csv(neo4j_data_path + filename, index=False)\n",
    "    if log:\n",
    "        _log_df(df)\n",
    "    return df\n",
    "\n",
    "\n",
    "def get_neo4j_data():\n",
    "    sotu_nodes = fetch_cached_df(\n",
    "        query_sotu_nodes,\n",
    "        {\n",
    "            'tax_ids': tax_ids,\n",
    "            'zoonotic_tax_ids': zoonotic_tax_ids,\n",
    "        },\n",
    "        'sotu_nodes.csv'\n",
    "    )\n",
    "    sotus = sotu_nodes['palmId'].unique().tolist()\n",
    "    sotu_msa_edges = fetch_cached_df(\n",
    "        query_sotu_msa_edges,\n",
    "        {\n",
    "            'tax_ids': tax_ids,\n",
    "            'sotus': sotus,\n",
    "        },\n",
    "        'sotu_msa_edges.csv'\n",
    "    )\n",
    "    sotu_nodes['taxId'] = sotu_nodes['taxId'].astype(int)\n",
    "\n",
    "    taxon_order_nodes = fetch_cached_df(\n",
    "        query_stat_taxon_order_nodes,\n",
    "        {\n",
    "            'sotus': sotus,\n",
    "            'host_class_tax_ids': host_class_tax_ids,\n",
    "        },\n",
    "        'taxon_order_nodes.csv',\n",
    "    )\n",
    "    taxon_order_nodes['taxId'] = taxon_order_nodes['taxId'].astype(int)\n",
    "    \n",
    "\n",
    "    taxon_nodes = fetch_cached_df(\n",
    "        query_stat_taxon_nodes,\n",
    "        {\n",
    "            'sotus': sotus,\n",
    "            'host_class_tax_ids': host_class_tax_ids,\n",
    "        },\n",
    "        'taxon_nodes.csv',\n",
    "    )\n",
    "    taxon_nodes['taxId'] = taxon_nodes['taxId'].astype(int)\n",
    "\n",
    "    has_host_order_stat_edges = fetch_cached_df(\n",
    "        query_has_host_order_stat,\n",
    "        {\n",
    "            'sotus': sotus,\n",
    "        },\n",
    "        'has_host_order_stat_edges.csv',\n",
    "    )\n",
    "\n",
    "    has_host_stat_edges = fetch_cached_df(\n",
    "        query_has_host_stat,\n",
    "        {\n",
    "            'sotus': sotus,\n",
    "        },\n",
    "        'has_host_stat_edges.csv',\n",
    "    )\n",
    "\n",
    "    return {\n",
    "        'sotu_nodes': sotu_nodes,\n",
    "        'sotu_msa_edges': sotu_msa_edges,\n",
    "        'taxon_order_nodes': taxon_order_nodes,\n",
    "        'taxon_nodes': taxon_nodes,\n",
    "        'has_host_stat_edges': has_host_stat_edges,\n",
    "        'has_host_order_stat_edges': has_host_order_stat_edges,\n",
    "    }\n",
    "\n",
    "neo4j_data = get_neo4j_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(742, 8)\n",
      "(2714, 7)\n",
      "(613, 7)\n",
      "(1852, 11)\n"
     ]
    }
   ],
   "source": [
    "print(neo4j_data['sotu_nodes'].shape)\n",
    "print(neo4j_data['sotu_msa_edges'].shape)\n",
    "print(neo4j_data['taxon_nodes'].shape)\n",
    "print(neo4j_data['has_host_order_stat_edges'].shape)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distribution investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "507\n",
      "742\n"
     ]
    }
   ],
   "source": [
    "print(len(neo4j_data['has_host_order_stat_edges']['sourceAppId'].unique()))\n",
    "print(len(neo4j_data['sotu_nodes']['appId'].unique()) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "145856     215\n",
      "291484      47\n",
      "1239567     31\n",
      "11277       25\n",
      "28875       20\n",
      "          ... \n",
      "37124        1\n",
      "909207       1\n",
      "11080        1\n",
      "356862       1\n",
      "1046251      1\n",
      "Name: taxId, Length: 181, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "unique_tax_orders =  neo4j_data['sotu_nodes']['taxId'].value_counts()\n",
    "print(unique_tax_orders)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique SOTU taxIds with primate assoc 70\n",
      "unique SOTU taxIds with isZoonotic 77\n",
      "SOTUs taxIds in both 38\n",
      "SOTUs taxIds unique to isZoonotic 39\n",
      "SOTUs taxIds unique to primate assoc 32\n",
      "\n",
      "unique SOTUs with primate assoc 128\n",
      "unique SOTUs with isZoonotic 477\n",
      "SOTUs in both 72\n",
      "SOTUs unique to isZoonotic 405\n",
      "SOTUs unique to primate assoc 56\n"
     ]
    }
   ],
   "source": [
    "# Issue: Check on unclassified: 12908 (percentIdentityFull < 0.2)\n",
    "\n",
    "\n",
    "\n",
    "has_primate_assoc = neo4j_data['has_host_order_stat_edges'].loc[neo4j_data['has_host_order_stat_edges']['targetAppId'] == 9443]['sourceAppId'].unique()\n",
    "has_primate_assoc_tax = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['appId'].isin(has_primate_assoc)]['taxId'].unique()\n",
    "has_zoonotic_potential = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['isZoonotic'] == True]['appId'].unique()\n",
    "has_zoonotic_potential_tax = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['appId'].isin(has_zoonotic_potential)]['taxId'].unique()\n",
    "\n",
    "\n",
    "print(f'unique SOTU taxIds with primate assoc {len(has_primate_assoc_tax)}')\n",
    "print(f'unique SOTU taxIds with isZoonotic {len(has_zoonotic_potential_tax)}')\n",
    "\n",
    "print(f'SOTUs taxIds in both {len(set(has_zoonotic_potential_tax) & set(has_primate_assoc_tax)) }')\n",
    "print(f'SOTUs taxIds unique to isZoonotic {len(set(has_zoonotic_potential_tax) - set(has_primate_assoc_tax)) }')\n",
    "print(f'SOTUs taxIds unique to primate assoc {len(set(has_primate_assoc_tax) - set(has_zoonotic_potential_tax)) }')\n",
    "print()\n",
    "\n",
    "# majority class, only 181 unique SOTU taxIds, taxId 145856 has 215 of 477 SOTUs\n",
    "\n",
    "print(f'unique SOTUs with primate assoc {len(has_primate_assoc)}')\n",
    "print(f'unique SOTUs with isZoonotic {len(has_zoonotic_potential)}')\n",
    "\n",
    "print(f'SOTUs in both {len(set(has_zoonotic_potential) & set(has_primate_assoc)) }')\n",
    "print(f'SOTUs unique to isZoonotic {len(set(has_zoonotic_potential) - set(has_primate_assoc)) }')\n",
    "print(f'SOTUs unique to primate assoc {len(set(has_primate_assoc) - set(has_zoonotic_potential)) }')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Community detection and node centrality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_gds_projection(neo4j_data):\n",
    "    graph_name = 'zoonotic'\n",
    "    nodes = pd.concat([\n",
    "        neo4j_data['sotu_nodes'][['nodeId', 'labels', 'taxId']],\n",
    "        # neo4j_data['taxon_order_nodes'][['nodeId', 'labels']],\n",
    "        # neo4j_data['taxon_nodes_missing_order'][['nodeId', 'labels']],\n",
    "    ])\n",
    "    relationships = pd.concat([\n",
    "        neo4j_data['sotu_msa_edges'][['sourceNodeId', 'targetNodeId', 'relationshipType', 'weight']],\n",
    "    ])\n",
    "\n",
    "    if gds.graph.exists(graph_name)['exists']:\n",
    "        gds.graph.drop(gds.graph.get(graph_name))\n",
    "\n",
    "    G = gds.alpha.graph.construct(\n",
    "        graph_name=graph_name,\n",
    "        nodes=nodes,\n",
    "        relationships=relationships,\n",
    "        concurrency=4,\n",
    "        undirected_relationship_types=['HAS_HOST_STAT'],\n",
    "    )\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_community_analysis(G):\n",
    "  communities = gds.labelPropagation.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    maxIterations=30,\n",
    "    seedProperty='taxId',\n",
    "  )\n",
    "  unique_communities = communities.communityId.unique()\n",
    "  community_counter = collections.Counter(communities.communityId)\n",
    "  print('LPA Unique communities:', len(unique_communities))\n",
    "  print('LPA Most common communities:', community_counter.most_common(10))\n",
    "\n",
    "  wcc = gds.wcc.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    threshold=None,\n",
    "    seedProperty='taxId'\n",
    "  )\n",
    "\n",
    "  unique_communities = wcc.componentId.unique()\n",
    "  community_counter = collections.Counter(wcc.componentId)\n",
    "  print('WCC Unique communities:', len(unique_communities))\n",
    "  print('WCC Most common communities:', community_counter.most_common(10))\n",
    "\n",
    "  page_ranks_sotu = gds.pageRank.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    maxIterations=100,\n",
    "  )\n",
    "  page_ranks_sotu = page_ranks_sotu.rename(columns={'score': 'pageRankSotu'})\n",
    "  page_ranks_sotu['pageRankSotu'] = page_ranks_sotu['pageRankSotu'].round(0)\n",
    "\n",
    "  louvain = gds.louvain.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    seedProperty='taxId',\n",
    "  )\n",
    "  unique_communities = louvain.communityId.unique()\n",
    "  community_counter = collections.Counter(louvain.communityId)\n",
    "  print('Lieden Unique communities:', len(unique_communities))\n",
    "  print('Lieden Most common communities:', community_counter.most_common(10))\n",
    "\n",
    "  return communities, wcc, page_ranks_sotu, louvain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LPA Unique communities: 110\n",
      "LPA Most common communities: [(145856, 215), (291484, 48), (11277, 47), (1239567, 37), (64300, 27), (28875, 20), (28344, 17), (147711, 17), (40058, 15), (11033, 15)]\n",
      "WCC Unique communities: 98\n",
      "WCC Most common communities: [(145856, 215), (291484, 45), (11072, 43), (11272, 41), (28875, 24), (40050, 23), (1239567, 22), (147711, 19), (1239566, 18), (11020, 16)]\n",
      "Lieden Unique communities: 80\n",
      "Lieden Most common communities: [(145856, 215), (1239574, 49), (11303, 48), (291484, 48), (12637, 46), (11036, 25), (1679172, 24), (28876, 24), (147712, 22), (2593991, 21)]\n"
     ]
    }
   ],
   "source": [
    "G = construct_gds_projection(neo4j_data)\n",
    "communities, wcc, page_ranks_sotu, louvain = run_community_analysis(G)\n",
    "G.drop()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Graphistry visualizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create node and relationship dataframes with full information\n",
    "def get_graphistry_df(\n",
    "        neo4j_data, communities, page_ranks_sotu\n",
    "    ):\n",
    "\n",
    "    neo4j_data['sotu_nodes']['displayLabel'] = neo4j_data['sotu_nodes']['appId']\n",
    "    neo4j_data['taxon_order_nodes']['displayLabel'] = neo4j_data['taxon_order_nodes']['appId']\n",
    "\n",
    "    nodes = pd.concat([\n",
    "        neo4j_data['sotu_nodes'],\n",
    "        neo4j_data['taxon_order_nodes'],\n",
    "    ])\n",
    "    nodes = nodes.merge(\n",
    "        communities,\n",
    "        left_on='nodeId',\n",
    "        right_on='nodeId',\n",
    "        how='left',\n",
    "    )\n",
    "    nodes = nodes.merge(\n",
    "        page_ranks_sotu,\n",
    "        left_on='nodeId',\n",
    "        right_on='nodeId',\n",
    "        how='left',\n",
    "    )\n",
    "\n",
    "    nodes['type'] = nodes['labels']\n",
    "\n",
    "    nodes = nodes[[\n",
    "        'appId', 'labels', 'type',\n",
    "        'taxId', 'isZoonotic', 'hostClass',\n",
    "        'taxRank', 'communityId',\n",
    "        'taxOrder', 'displayLabel',\n",
    "    ]].astype(str)\n",
    "\n",
    "    # Encode communityId to use default color pallette\n",
    "    nodes['communityId'] = nodes['communityId'].astype('float64')\n",
    "    nodes['communityId'] = nodes['communityId'].fillna(-1)\n",
    "    nodes['communityId'] = nodes['communityId'].astype('int32')\n",
    "    labels = nodes['communityId'].unique()\n",
    "    mapping = {label: i for i, label in enumerate(labels)}\n",
    "    nodes['communityId'] =  nodes['communityId'].replace(mapping)\n",
    "    nodes['communityColorCodes'] = nodes['communityId'].mod(11)\n",
    "\n",
    "    ## Aggregate statCoverage for nodes using average of edges\n",
    "    agg_source = neo4j_data['has_host_stat_edges'].groupby('sourceAppId').agg({'statCoverage': 'mean'})\n",
    "    agg_source['statCoverage'] = agg_source['statCoverage']\n",
    "    agg_target = neo4j_data['has_host_stat_edges'].groupby('targetAppId').agg({'statCoverage': 'mean'})\n",
    "    agg_target['statCoverage'] = agg_target['statCoverage']\n",
    "    aggs = pd.concat([agg_source, agg_target])\n",
    "    aggs = aggs.reset_index()\n",
    "    aggs['statCoverageInt'] = round(aggs['statCoverage'] * 100)\n",
    "    aggs['statCoverage'] = aggs['statCoverage'].round(1)\n",
    "    aggs['statCoverageInt'] = aggs['statCoverageInt'].round(0).astype('int32')\n",
    "    aggs = aggs.rename({'index': 'appId'}, axis='columns')\n",
    "    nodes = nodes.merge(\n",
    "        aggs,\n",
    "        left_on='appId',\n",
    "        right_on='appId',\n",
    "        how='left',\n",
    "    )\n",
    "\n",
    "    # Add host counts for palmprints\n",
    "    stat_hosts_excl_mixed = neo4j_data['has_host_stat_edges'][neo4j_data['has_host_stat_edges']['targetAppId'] != 12908]\n",
    "    host_counts = stat_hosts_excl_mixed.groupby('sourceAppId').size().groupby(level=0).max()\n",
    "    # host_counts = neo4j_data['has_host_stat_edges'].groupby('sourceAppId').size().groupby(level=0).max()\n",
    "    host_counts = host_counts.reset_index()\n",
    "    host_counts = host_counts.rename({0: 'hostCount'}, axis='columns')\n",
    "    host_counts['hostCountNormalized'] = round((host_counts['hostCount'] / host_counts['hostCount'].max()) * 100, -1).astype('int32')\n",
    "    nodes = nodes.merge(\n",
    "        host_counts,\n",
    "        left_on='appId',\n",
    "        right_on='sourceAppId',\n",
    "        how='left',\n",
    "    )\n",
    "\n",
    "    relationships = pd.concat([\n",
    "        neo4j_data['sotu_msa_edges'],\n",
    "        neo4j_data['has_host_stat_edges'],\n",
    "    ])\n",
    "    \n",
    "    relationships['targetAppId'] = relationships['targetAppId'].astype(str)\n",
    "    relationships['sourceAppId'] = relationships['sourceAppId'].astype(str)\n",
    "\n",
    "    relationships['weight'] = relationships['weight'].astype(float)\n",
    "    relationships['weightInt'] = round(relationships['weight'] * 100, -1).astype('int32')\n",
    "    relationships = relationships[[\n",
    "        'sourceAppId', 'targetAppId', 'relationshipType', \n",
    "        'weight', 'weightInt',\n",
    "    ]].astype(str)\n",
    "    relationships['weightInt'] = relationships['weightInt'].astype('int32')\n",
    "    return nodes, relationships\n",
    "\n",
    "\n",
    "\n",
    "def load_or_create_graphistry_df(node_filename, relationship_filename, use_cache=False):\n",
    "    if use_cache and os.path.exists(graphistry_data_path + node_filename) \\\n",
    "            and os.path.exists(graphistry_data_path + relationship_filename):\n",
    "        nodes = pd.read_csv(graphistry_data_path + node_filename)\n",
    "        edges = pd.read_csv(graphistry_data_path + relationship_filename)\n",
    "    else:\n",
    "        neo4j_data = get_neo4j_data()\n",
    "        G = construct_gds_projection(neo4j_data)\n",
    "        communities, wcc, page_ranks_sotu = run_community_analysis(G)\n",
    "        # binary_pvals, multi_pvals = run_pval_analysis(neo4j_data)\n",
    "        nodes, edges = get_graphistry_df(\n",
    "            neo4j_data, communities, page_ranks_sotu\n",
    "        )\n",
    "        nodes.to_csv(graphistry_data_path + node_filename, index=False)\n",
    "        edges.to_csv(graphistry_data_path + relationship_filename, index=False)\n",
    "    return nodes, edges\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'construct_gds_projection' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[78], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m nodes, relationships \u001b[39m=\u001b[39m load_or_create_graphistry_df(\u001b[39m'\u001b[39m\u001b[39mnodes.csv\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39medges.csv\u001b[39m\u001b[39m'\u001b[39m)\n",
      "Cell \u001b[0;32mIn[77], line 102\u001b[0m, in \u001b[0;36mload_or_create_graphistry_df\u001b[0;34m(node_filename, relationship_filename, use_cache)\u001b[0m\n\u001b[1;32m    100\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    101\u001b[0m     neo4j_data \u001b[39m=\u001b[39m get_neo4j_data()\n\u001b[0;32m--> 102\u001b[0m     G \u001b[39m=\u001b[39m construct_gds_projection(neo4j_data)\n\u001b[1;32m    103\u001b[0m     communities, wcc, page_ranks_sotu \u001b[39m=\u001b[39m run_community_analysis(G)\n\u001b[1;32m    104\u001b[0m     \u001b[39m# binary_pvals, multi_pvals = run_pval_analysis(neo4j_data)\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'construct_gds_projection' is not defined"
     ]
    }
   ],
   "source": [
    "nodes, relationships = load_or_create_graphistry_df('nodes.csv', 'edges.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_color_pallete = [\n",
    "    \"rgb(166, 206, 227)\",\n",
    "    \"rgb(31, 120, 180)\",\n",
    "    \"rgb(178, 223, 138)\",\n",
    "    \"rgb(51, 160, 44)\",\n",
    "    \"rgb(251, 154, 153)\",\n",
    "    \"rgb(227, 26, 28)\",\n",
    "    \"rgb(253, 191, 111)\",\n",
    "    \"rgb(255, 127, 0)\",\n",
    "    \"rgb(202, 178, 214)\",\n",
    "    \"rgb(106, 61, 154)\",\n",
    "    \"rgb(255, 255, 153)\", \n",
    "    \"#ffffff\",\n",
    "]\n",
    "\n",
    "categorical_colors = {}\n",
    "\n",
    "for communityId in nodes.communityColorCodes.unique():\n",
    "    categorical_colors[str(communityId)] = alt_color_pallete[communityId % len(alt_color_pallete)]\n",
    "\n",
    "categorical_colors[11] = \"#808080\"\n",
    "\n",
    "# for i, row in nodes.iterrows():\n",
    "#     if row['hostCount'] > 1:\n",
    "#         nodes.at[i,'communityColorCodes'] = 11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <iframe id=\"d590395b-1861-4eb7-9a5f-d75a83113dc9\" src=\"https://hub.graphistry.com/graph/graph.html?dataset=c4c26d29d2884edd93e20823a5afae3b&type=arrow&viztoken=81a29871-33b6-4306-89f2-201efa001fa3&usertag=66cc8c0c-pygraphistry-0.29.2&splashAfter=1698428208&info=True&play=2000&menu=True&showArrows=True&edgeOpacity=0.25&pointOpacity=1.0&linLog=True&compactLayout=True&strongGravity=True&dissuadeHubs=True&edgeInfluence=5&showLabelPropertiesOnHover=True&pointsOfInterestMax=15&backgroundColor=black\"\n",
       "                    allowfullscreen=\"true\" webkitallowfullscreen=\"true\" mozallowfullscreen=\"true\"\n",
       "                    oallowfullscreen=\"true\" msallowfullscreen=\"true\"\n",
       "                    style=\"width:100%; height:500px; border: 1px solid #DDD; overflow: hidden\"\n",
       "                    \n",
       "            >\n",
       "            </iframe>\n",
       "        \n",
       "            <script>\n",
       "                try {\n",
       "                  $(\"#d590395b-1861-4eb7-9a5f-d75a83113dc9\").bind('mousewheel', function(e) { e.preventDefault(); });\n",
       "                } catch (e) { console.error('exn catching scroll', e); }\n",
       "            </script>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g = graphistry.bind()\n",
    "\n",
    "g = g.bind(\n",
    "    source='sourceAppId',\n",
    "    destination='targetAppId',\n",
    "    edge_weight='weight',\n",
    ").edges(relationships)\n",
    "\n",
    "g = g.bind(\n",
    "    node='appId',\n",
    "    point_label='displayLabel',\n",
    "    point_size=None,\n",
    "    # point_size='statCoverageInt',\n",
    ").nodes(nodes)\n",
    "\n",
    "\n",
    "params = {\n",
    "        'play': 2000,\n",
    "        'menu': True, \n",
    "        'info': True,\n",
    "        'showArrows': True,\n",
    "        # 'pointSize': 2.0, \n",
    "        # 'edgeCurvature': 0.5,\n",
    "        'edgeOpacity': 0.25, \n",
    "        'pointOpacity': 1.0,\n",
    "        # 'lockedX': False, 'lockedY': False, 'lockedR': False,\n",
    "        'linLog': True, \n",
    "        'compactLayout': True,\n",
    "        'strongGravity': True,\n",
    "        'dissuadeHubs': True,\n",
    "        'edgeInfluence': 5,\n",
    "        # 'precisionVsSpeed': 0, 'gravity': 1.0, 'scalingRatio': 1.0,\n",
    "        # 'showLabels': True, 'showLabelOnHover': True,\n",
    "        # 'showPointsOfInterest': True, 'showPointsOfInterestLabel': True, \n",
    "        'showLabelPropertiesOnHover': True,\n",
    "        'pointsOfInterestMax': 15,\n",
    "        'backgroundColor': 'black',\n",
    "      }\n",
    "\n",
    "g = g.settings(url_params=params)\n",
    "\n",
    "\n",
    "g = g.addStyle(\n",
    "    bg={\n",
    "        'color': 'white',\n",
    "})\n",
    "\n",
    "# g = g.encode_point_color(\n",
    "#     'communityIds',\n",
    "#     categorical_mapping=categorical_colors,\n",
    "#     default_mapping='grey', \n",
    "# )\n",
    "\n",
    "g = g.encode_point_color(\n",
    "    'communityColorCodes',\n",
    "    categorical_mapping=categorical_colors,\n",
    ")\n",
    "\n",
    "g.plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(urllib.parse.urlencode(params))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notes:\n",
    "\n",
    "- General takeaways:\n",
    "    - Clustering SOTUs using weight mostly correlates to taxOrder, with some subcommunities within communities. Reovirales doesn't cluster well\n",
    "- \n",
    "\n",
    "- SOTU only, default community detection\n",
    "    - 742 nodes, 2714 edges\n",
    "    - LPA Unique communities: 438\n",
    "    - LPA Most common communities: [(8170199, 45), (7858962, 39), (7743098, 27), (7935516, 20), (7999668, 15), (7971675, 13), (7714332, 12), (8100725, 12), (7994679, 12), (8017721, 10)]\n",
    "    - WCC Unique communities: 386\n",
    "    - WCC Most common communities: [(431, 86), (2, 43), (0, 40), (5, 23), (34, 17), (48, 16), (79, 16), (41, 15), (1, 14), (106, 14)]\n",
    "    - Lieden Unique communities: 80\n",
    "    - Lieden Most common communities: [(145856, 215), (1239574, 49), (11303, 48), (291484, 48), (12637, 46), (11036, 25), (1679172, 24), (28876, 24), (147712, 22), (2593991, 21)]\n",
    "    - 356 stand alone nodes, 386 remain after hiding stand alone nodes\n",
    "    - standalone: zoonotic: 245/477, non-zoonotic: 141/265\n",
    "    - communities are mostly homogenous but contain some other colors (?)\n",
    "    - non-zoonotic are mixed in med-small communities, have med-small node degree\n",
    "    - zoonotic make up entire largest connected community\n",
    "    - \n",
    "\n",
    "- SOTU only, seeded community detection\n",
    "    - 739 nodes, 2570 edges\n",
    "    - LP Unique communities: 110\n",
    "    - LP Most common communities: [(145856, 215), (291484, 48), (11277, 47), (1239567, 37), (64300, 27), (28875, 20), (28344, 17), (147711, 17), (40058, 15), (11033, 15)]\n",
    "    - WCC Unique communities: 98\n",
    "    - WCC Most common communities: [(145856, 215), (291484, 45), (11072, 43), (11272, 41), (28875, 24), (40050, 23), (1239567, 22), (147711, 19), (1239566, 18), (11020, 16)]\n",
    "    - Lieden Unique communities: 80\n",
    "    - Lieden Most common communities: [(145856, 215), (1239574, 49), (11303, 48), (291484, 48), (12637, 46), (11036, 25), (1679172, 24), (28876, 24), (147712, 22), (2593991, 21)]\n",
    "    - 356 stand alone nodes, 383 remain after hiding stand alone nodes\n",
    "    - standalone: zoonotic: 242/477, non-zoonotic: 141/265\n",
    "    - communities are more homogenous than the default\n",
    "    - taxOrder and taxId corresponds nicely with clusters\n",
    "    - largest community is even larger\n",
    "    - overall zoonotic/non-zoonotic appears the same\n",
    "    - non-zoonotic exist in larger community now\n",
    "    - low-weight edges seem to involve long-range\n",
    "\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rnalab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "448a416c49845ffbddc886562179757704d684dae284cc536cfa1eca10a7a7d0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
