{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Zoonotic Potential"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [*] remember to turn of EC2 instance\n",
    "\n",
    "\n",
    "### Planning\n",
    "\n",
    "- measure host abundance for each zoonotic SOTU\n",
    "- measure host abundance for each non-zoonotic SOTU\n",
    "- measure differential expression between positive and negative sets\n",
    "- statistical test of differential AGS with FGS of zoonotic potential (Virus set enrichment analysis)\n",
    "\n",
    "\n",
    "### Done\n",
    "\n",
    "- [x] parse virus dataset\n",
    "- [x] find number of genuses with differing in zoonotic values\n",
    "- [x] find rank distribution of has_potential_host\n",
    "    - total: 438167, species: 414190, genus: 24, family: 49\n",
    "- [x] manually patch tax_ids for all species\n",
    "- [x] find number of virus species with associated SOTU\n",
    "    - match genus taxons via list\n",
    "- [x] create projection of SOTUs and sequence alignments using zoonotic viruses\n",
    "- [x] setup graphistry visualization\n",
    "- [x] clustering analysis on SOTUs\n",
    "    - [x] setup community detection with seeded taxId\n",
    "    - [x] WCC\n",
    "    - [x] Louvain and lieden\n",
    "- [x] Include STAT hosts into projection \n",
    "- [x] understand discrepancy between zoonotic label and human host association\n",
    "    - goal: may need to re-examine using dataset for FGS?\n",
    "    - may be able to create our own zoonotic dataset using STAT?\n",
    "    - create a functional gene set (FGS) using entirety of the network of viruses in humans? if there is strong agreement\n",
    "- [x] update Graphistry with stat hosts (order level only)\n",
    "- [x] decide on using community or individual virus order/species\n",
    "    - find differentially expression of hosts (AGS) between positive and negative sets\n",
    "    - compare with FGS from entire network with specific hosts?\n",
    "- [x] sanity check clustering threshold at 80 - 90 %\n",
    "    - Seq_alignment\n",
    "        - max(r.percentIdentity): 0.951\n",
    "        - avg(r.percentIdentity): 0.584658444611331\n",
    "- [x] for each species find abundance of host taxons\n",
    "    - [x] remap host abundance by class [mammals, fish, plants, fungii, parasite]\n",
    "    - [x] find taxids for each\n",
    "    - store in CSV/table with each species as a row (index), column for zoonotic potential, other columns as classes, \n",
    "- [x] issue:\n",
    "    - sotu exists in sotu_nodes: u29006\n",
    "    - sotu doesn't exist in has_host_order_stat edges\n",
    "    - sotu should have edge: 548681\n",
    "    - taxon exists in taxon nodes: 548681\n",
    "- [x] Issue: x SOTUs have no associated host? could be related to threshold\n",
    "    - seems to be related to metagenome samples with weak STAT associations\n",
    "    - could consider ignoring unclassified sequence labels: 12908\n",
    "    - kmerPercentFull is comparing kmers to total spots?\n",
    "    - q: how does this impact hostClass distributions\n",
    "- [x] statistics on STAT host range\n",
    "    - compare variance within zoonotic and non-zoonotic\n",
    "        - log intensity-ratio\n",
    "    - compare differential change between zoonotic and non-zoonotic\n",
    "        - check if using DSEQ2 works\n",
    "        - rank-sum test \n",
    "        - https://nbisweden.github.io/excelerate-scRNAseq/session-de/session-de-methods.html\n",
    "        - find pvals, Fold-change, log-of-the-fold-changes for each class of zoonotics\n",
    "        - rank host range by significance and abundance (Sign(FC) \\* -log_10(pvalue))\n",
    "        - find significantly upregulated or down regulated traits to create functional class (gene set)\n",
    "- [x] Investigate STAT vs zoonotic label:\n",
    "    - check specific counts of zoonotic SOTUs with primate hosts\n",
    "    - check specific counts of non-zoonotc with primate hosts\n",
    "- [x] investigate STAT threshold:\n",
    "    - vary threshold to use percentIdentity >= 0.2\n",
    "    - what does it mean if zoonotic viruses are dependent on metagenome samples, namely upregulated in bacteria\n",
    "    - why do SOTUs appear in sotu_nodes without and relationships?\n",
    "-\n",
    "\n",
    "### TODO\n",
    "\n",
    "- variational analysis:\n",
    "    - modify code to run all steps for different STAT threshold\n",
    "    - save data in folders\n",
    "    - create panel of plots:\n",
    "        - Mammals and Bacteria (other classes as well)\n",
    "        - vary STAT threshold\n",
    "        - create scatter bar plots for z_0, z_1, z_t, z_r\n",
    "- \n",
    "\n",
    "- investigate metagenome samples:\n",
    "    - coarser: use metagenomic label instead of STAT\n",
    "    - granular: use taxOrder instead of hostClass\n",
    "    - post in slack\n",
    "    - [*] investigate hostClass labels:\n",
    "        - Move to seperate query and dataframe, merge after (fix bug with duplicate rows)\n",
    "        - Test: Add metagenomic class for unclassified tax_id based on metadata labe?\n",
    "        - manually review other, bacteriam etc\n",
    "-\n",
    "\n",
    "\n",
    "- -[*] look into how to increase dataset coverage\n",
    "    - check for genbank ids in SQL\n",
    "    - Many SOTUs have no host, likely related to threshold on metagenome samples, thresholding on percentIdentity seems sufficent, 0.2\n",
    "-\n",
    "\n",
    "- for each {community, species} measure network topology metric of sotus and full trait hierachy\n",
    "    - do I need to use a community of SOTUs?\n",
    "    - Modularity, etc\n",
    "    - spectrum\n",
    "\n",
    "-  topological link prediction\n",
    "\n",
    "- compute full overlap (jaccard similarity) of all pathogen interactions and STAT hosts?\n",
    "\n",
    "- Decide on including tissues\n",
    "    - get list of unique SRAs\n",
    "    - [*] check overlap of SRA runs in SQL tissue table \n",
    "    - for each {community, species} find tissue abundance by class\n",
    "\n",
    "- Add global biotic interaction data to graph\n",
    "    - https://www.globalbioticinteractions.org/?interactionType=interactsWith&sourceTaxon=NCBI%3A11272\n",
    "    - https://www.globalbioticinteractions.org/data#interaction-data-indexes\n",
    "    - add isHumanPathogen value? or GLOBI_PATHOGEN_OF\n",
    "    - can download all interactions as NEO4j graph db\n",
    "    - deeper investigation of known pathogens, STAT host associations, Host metadata labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The dotenv extension is already loaded. To reload it, use:\n",
      "  %reload_ext dotenv\n"
     ]
    }
   ],
   "source": [
    "# Imports and config for the zoonotic-potential notebook\n",
    "\n",
    "## Notebook config\n",
    "import sys\n",
    "if '../' not in sys.path:\n",
    "    sys.path.append(\"../\")\n",
    "%load_ext dotenv\n",
    "%reload_ext dotenv\n",
    "%dotenv\n",
    "\n",
    "import collections\n",
    "import os\n",
    "import urllib.parse\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import graphistry\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "\n",
    "from datasources.neo4j import gds\n",
    "from queries import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<graphdatascience.graph_data_science.GraphDataScience object at 0x13d4d15d0>\n"
     ]
    }
   ],
   "source": [
    "# Configs\n",
    "\n",
    "base_data_path = './zoonosis_data/'\n",
    "virus_dataset = base_data_path + 'trefle/'\n",
    "neo4j_data_path = base_data_path + 'neo4j/'\n",
    "graphistry_data_path = base_data_path + 'graphistry/'\n",
    "outputs = base_data_path + 'outputs/' \n",
    "\n",
    "graphistry.register(\n",
    "    api=3,\n",
    "    username=os.getenv('GRAPHISTRY_USERNAME'),\n",
    "    password=os.getenv('GRAPHISTRY_PASSWORD'),\n",
    ")\n",
    "\n",
    "print(gds)\n",
    "\n",
    "df = pd.read_csv(virus_dataset + 'viruses.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fill in tax_ids (only run once)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_names = df['vVirusNameCorrected'].unique()\n",
    "\n",
    "def parse_scientific_name(name):\n",
    "    \"\"\"\n",
    "    Parse a scientific name into its components.\n",
    "    \"\"\"\n",
    "    name = name.replace('_', ' ')\n",
    "    return name\n",
    "\n",
    "with open( virus_dataset + 'species_names.txt', 'w') as f:\n",
    "    for line in unique_names:\n",
    "        f.write(f\"{parse_scientific_name(line)}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "cat ./zoonosis_data/trefle/species_names.txt | taxonkit name2taxid  >> ./zoonosis_data/trefle/species_taxids.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                name    taxid\n",
      "0               Adelaide River virus  31612.0\n",
      "1           Adeno-associated virus-1      NaN\n",
      "2           Adeno-associated virus-2      NaN\n",
      "3           Adeno-associated virus-5      NaN\n",
      "4  African green monkey polyomavirus  12480.0\n",
      "(586, 2)\n",
      "269\n"
     ]
    }
   ],
   "source": [
    "df2 = pd.read_csv(virus_dataset + 'species_taxids.tsv', sep='\\t', header=None, names=['name', 'taxid'])\n",
    "print(df2.head())\n",
    "print(df2.shape)\n",
    "print(df2['taxid'].isna().sum())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load zoonotic dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22,)\n",
      "(586, 24)\n",
      "(564, 24)\n",
      "\n",
      "0    378\n",
      "1    186\n",
      "Name: IsZoonotic, dtype: int64\n",
      "0    458\n",
      "1    106\n",
      "Name: IsZoonotic.stringent, dtype: int64\n",
      "0    557\n",
      "1      7\n",
      "Name: ReverseZoonoses, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/vl/wsxpm_412lgcfm03dqd31qmh0000gn/T/ipykernel_16085/3246926767.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(int)\n",
      "/var/folders/vl/wsxpm_412lgcfm03dqd31qmh0000gn/T/ipykernel_16085/3246926767.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(str)\n"
     ]
    }
   ],
   "source": [
    "# Manually fill in any unresolved taxids\n",
    "df2 = pd.read_csv(virus_dataset + 'species_taxids_patched.tsv', sep='\\t', header=None, names=['name', 'taxid'])\n",
    "df2['name'] = df2['name'].str.replace(' ', '_')\n",
    "missing_taxid = df2[df2['taxid'].isna()]['name']\n",
    "\n",
    "print(missing_taxid.shape)\n",
    "\n",
    "merged = pd.merge(df, df2, left_on='vVirusNameCorrected', right_on='name', how='left')\n",
    "merged = merged.drop(columns=['name'])\n",
    "merged = merged.rename(columns={'taxid': 'TaxID'})\n",
    "merged_dropna = merged.dropna(subset=['TaxID'])\n",
    "merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(int)\n",
    "merged_dropna['TaxID'] = merged_dropna['TaxID'].astype(str)\n",
    "\n",
    "print(merged.shape)\n",
    "print(merged_dropna.shape)\n",
    "print()\n",
    "\n",
    "print(merged_dropna['IsZoonotic'].value_counts())\n",
    "print(merged_dropna['IsZoonotic.stringent'].value_counts())\n",
    "print(merged_dropna['ReverseZoonoses'].value_counts())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Neo4j queries"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Dataset Serratus overlap QA\n",
    "\n",
    "Mapping tax_id to virus name:\n",
    "- 586 virus in original dataset, 564 after parsing name to ncbi taxid, couldn't find 22 viruses in NCBI taxonomy. Of these 564, 378 are not zoonotic, 186 are zoonotic\n",
    "\n",
    "Mapping SOTUs to virus tax_id \n",
    "- Direct query of virus taxons from the dataset to mapped SOTUs results in coverage of 181 virus taxons (383 have no direct SOTUs). From these 181 virus taxons, there are 742 SOTU associations, max 215, mean ~4, std ~17. All of the SOTUs potential tax labels have strong percent identity and palmrint coverage (>= .95), i didn't account for e-values.\n",
    "- Including all children and any parents with 'species' level rank from the virus taxons in the dataset does not increase the number of mapped virus taxons or SOTUs. \n",
    "\n",
    "```\n",
    "v1:\n",
    "sotu_nodes:  (742, 8)\n",
    "sotu_msa_edges:  (2714, 7)\n",
    "taxon_order_nodes:  (187, 7)\n",
    "has_host_order_stat_edges:  (1346, 10)\n",
    "\n",
    "v2 [threshold on STAT @ 0.2]:\n",
    "sotu_nodes:  (742, 8)\n",
    "sotu_msa_edges:  (2714, 7)\n",
    "taxon_order_nodes:  (281, 7)\n",
    "has_host_order_stat_edges:  (2950, 10)\n",
    "\n",
    "v3 [threshold on STAT @ 0.5]:\n",
    "sotu_nodes:  (742, 8)\n",
    "sotu_msa_edges:  (2714, 7)\n",
    "taxon_order_nodes:  (231, 7)\n",
    "has_host_order_stat_edges:  (1642, 10)\n",
    "\n",
    "v4 [threshold on STAT @ 0.9]\n",
    "sotu_nodes:  (742, 8)\n",
    "sotu_msa_edges:  (2714, 7)\n",
    "taxon_order_nodes:  (167, 7)\n",
    "has_host_order_stat_edges:  (877, 10)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_sotu_nodes = \"\"\"\n",
    "    MATCH (a:SOTU)-[:HAS_POTENTIAL_TAXON]->(b:Taxon)\n",
    "    WHERE b.taxId in $tax_ids\n",
    "    RETURN\n",
    "        id(a) as nodeId,\n",
    "        a.palmId as appId,\n",
    "        a.palmId as palmId,\n",
    "        labels(a) as labels,\n",
    "        b.taxId as taxId,\n",
    "        b.rank as taxRank,\n",
    "        b.taxOrder as taxOrder,\n",
    "        CASE WHEN b.taxId in $zoonotic_tax_ids THEN True ELSE False END AS isZoonotic\n",
    "\"\"\"\n",
    "\n",
    "query_sotu_msa_edges = \"\"\"\n",
    "    MATCH (a:SOTU)-[r:SEQUENCE_ALIGNMENT]->(b:SOTU)\n",
    "    WHERE a.palmId in $sotus\n",
    "    AND b.palmId in $sotus\n",
    "    RETURN\n",
    "        id(a) as sourceNodeId,\n",
    "        a.palmId as sourceAppId,\n",
    "        id(b) as targetNodeId,\n",
    "        b.palmId as targetAppId,\n",
    "        'SEQUENCE_ALIGNMENT' as relationshipType,\n",
    "        r.percentIdentity as percentIdentity,\n",
    "        r.percentIdentity as weight\n",
    "\"\"\"\n",
    "\n",
    "query_stat_taxon_order_nodes = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        AND q.percentIdentity >= $stat_threshold\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        AND q.percentIdentity >= $stat_threshold\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "    }\n",
    "    WITH t, u\n",
    "    RETURN\n",
    "        id(t) as nodeId,\n",
    "        t.taxId as appId,\n",
    "        t.taxId as taxId,\n",
    "        labels(t) as labels,\n",
    "        t.rank as taxRank,\n",
    "        t.taxOrder as taxOrder,\n",
    "        CASE WHEN u.taxId = '40674' THEN 'Mammal' ELSE \n",
    "            CASE WHEN u.taxId = '33090' THEN 'Plant' ELSE \n",
    "                CASE WHEN u.taxId = '4751' THEN 'Fungi' ELSE \n",
    "                    CASE When u.taxId = '33630' THEN 'Parasite' ELSE \n",
    "                        CASE WHEN u.taxId = '2' THEN 'Bacteria' ELSE \n",
    "                            CASE WHEN u.taxId = '10239' THEN 'Virus' ELSE \n",
    "                                CASE WHEN u.taxId = '6656' THEN 'Anthropod' ELSE \n",
    "                                    CASE WHEN u.taxId = '7742' THEN 'Fish-like' ELSE 'Other'\n",
    "                                    END\n",
    "                                END\n",
    "                            END\n",
    "                        END\n",
    "                    END\n",
    "                END\n",
    "            END\n",
    "        END AS hostClass\n",
    "\"\"\"\n",
    "\n",
    "query_has_host_order_stat = '''\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        AND q.percentIdentity >= $stat_threshold\n",
    "        RETURN p, t, r, q\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->()-[:HAS_PARENT*0..]->(t:Taxon {rank: 'order'})\n",
    "        WHERE p.palmId in $sotus\n",
    "        AND q.percentIdentity >= $stat_threshold\n",
    "        RETURN p, t, r, q\n",
    "    }\n",
    "    WITH p, t, r, q\n",
    "    RETURN\n",
    "        id(p) as sourceNodeId,\n",
    "        p.palmId as sourceAppId,\n",
    "        id(t) as targetNodeId,\n",
    "        t.taxId as targetAppId,\n",
    "        //CASE WHEN percentIdentityFull >= 0.2 THEN id(t) ELSE 8765758 END AS targetNodeId,\n",
    "        //CASE WHEN percentIdentityFull >= 0.2 THEN t.taxId ELSE '12908' END AS targetAppId,\n",
    "        'HAS_HOST_STAT' as relationshipType,\n",
    "        count(*) AS count,\n",
    "        avg(r.percentIdentity) as avgPercentIdentityPalmprint,\n",
    "        avg(q.percentIdentity) as avgPercentIdentityStatKmers,\n",
    "        avg(q.percentIdentityFull) as avgPercentIdentityStatSpots,\n",
    "        avg(q.percentIdentity) * avg(r.percentIdentity) as weight\n",
    "'''\n",
    "\n",
    "query_host_class_label_map = '''\n",
    "    CALL {\n",
    "        MATCH (t:Taxon)\n",
    "        WHERE t.taxId in $host_tax_ids\n",
    "        OPTIONAL MATCH (t:Taxon)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE t.taxId in $host_tax_ids\n",
    "        AND u.taxId in $host_class_tax_ids\n",
    "        //OPTIONAL MATCH (t:Taxon)\n",
    "        //WHERE t.taxId in $host_class_tax_ids AND t.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "    }\n",
    "    WITH t, u\n",
    "    RETURN\n",
    "        t.taxId as taxId,\n",
    "        CASE WHEN u.taxId = '40674' THEN 'Mammal' ELSE \n",
    "            CASE WHEN u.taxId = '33090' THEN 'Plant' ELSE \n",
    "                CASE WHEN u.taxId = '4751' THEN 'Fungi' ELSE \n",
    "                    CASE WHEN u.taxId = '12908' THEN 'Metagenome' ELSE\n",
    "                        CASE When u.taxId = '33630' THEN 'Parasite' ELSE \n",
    "                            CASE WHEN u.taxId = '2' THEN 'Bacteria' ELSE \n",
    "                                CASE WHEN u.taxId = '10239' THEN 'Virus' ELSE \n",
    "                                    CASE WHEN u.taxId = '6656' THEN 'Anthropod' ELSE \n",
    "                                        CASE WHEN u.taxId = '7742' THEN 'Fish-like' ELSE \n",
    "                                            'Other'\n",
    "                                        END\n",
    "                                    END\n",
    "                                END\n",
    "                            END\n",
    "                        END\n",
    "                    END\n",
    "                END\n",
    "            END\n",
    "        END AS hostClass\n",
    "'''\n",
    "\n",
    "\n",
    "query_has_host_metadata_label = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, q\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, q\n",
    "    }\n",
    "    WITH p, t, r, q, avg(q.percentIdentityFull) as percentIdentityFull\n",
    "    RETURN\n",
    "        id(p) as sourceNodeId,\n",
    "        p.palmId as sourceAppId,\n",
    "        id(t) as targetNodeId,\n",
    "        t.taxId as targetAppId,\n",
    "        'HAS_HOST_METADATA_LABEL' as relationshipType,\n",
    "        count(*) AS count,\n",
    "        avg(q.percentIdentityFull) as avgPercentIdentityStatSpots,\n",
    "        1 as weight\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "### QA queries\n",
    "\n",
    "query_stat_taxon_order_nodes = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        OPTIONAL MATCH (t)-[:HAS_PARENT*0..]->(u:Taxon)\n",
    "        WHERE u.taxId in $host_class_tax_ids\n",
    "        RETURN t, u\n",
    "    }\n",
    "    WITH t, u\n",
    "    RETURN\n",
    "        id(t) as nodeId,\n",
    "        t.taxId as appId,\n",
    "        t.taxId as taxId,\n",
    "        labels(t) as labels,\n",
    "        t.rank as taxRank,\n",
    "        t.taxOrder as taxOrder,\n",
    "        CASE WHEN u.taxId = '40674' THEN 'Mammal' ELSE \n",
    "            CASE WHEN u.taxId = '33090' THEN 'Plant' ELSE \n",
    "                CASE WHEN u.taxId = '4751' THEN 'Fungi' ELSE \n",
    "                    CASE When u.taxId = '33630' THEN 'Parasite' ELSE \n",
    "                        CASE WHEN u.taxId = '2' THEN 'Bacteria' ELSE \n",
    "                            CASE WHEN u.taxId = '10239' THEN 'Virus' ELSE \n",
    "                                CASE WHEN u.taxId = '6656' THEN 'Anthropod' ELSE \n",
    "                                    CASE WHEN u.taxId = '7742' THEN 'Fish-like' ELSE 'Other'\n",
    "                                    END\n",
    "                                END\n",
    "                            END\n",
    "                        END\n",
    "                    END\n",
    "                END\n",
    "            END\n",
    "        END AS hostClass\n",
    "\"\"\"\n",
    "\n",
    "query_has_host_stat = '''\n",
    "    CALL {\n",
    "        MATCH (p:SOTU)<-[:HAS_SOTU]-(:Palmprint)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, q\n",
    "        UNION\n",
    "        MATCH (p:SOTU)<-[r:HAS_PALMPRINT]-(s:SRA)\n",
    "            -[q:HAS_HOST_STAT]->(t:Taxon)\n",
    "        WHERE p.palmId in $sotus\n",
    "        RETURN p, t, r, s, q\n",
    "    }\n",
    "    WITH p, t, r, s, q, avg(q.percentIdentityFull) as percentIdentityFull\n",
    "    RETURN\n",
    "        id(p) as sourceNodeId,\n",
    "        p.palmId as sourceAppId,\n",
    "        CASE WHEN percentIdentityFull >= 0.2 THEN id(t) ELSE 8765758 END AS targetNodeId,\n",
    "        CASE WHEN percentIdentityFull >= 0.2 THEN t.taxId ELSE '12908' END AS targetAppId,\n",
    "        'HAS_HOST_STAT' as relationshipType,\n",
    "        count(*) AS directAssociations,\n",
    "        count(*) AS count,\n",
    "        avg(r.percentIdentity) as avgPercentIdentityPalmprint,\n",
    "        avg(q.percentIdentity) as avgPercentIdentityStatKmers,\n",
    "        avg(q.percentIdentityFull) as avgPercentIdentityStatSpots,\n",
    "        avg(q.percentIdentity) * avg(r.percentIdentity) as weight\n",
    "'''\n",
    "\n",
    "query_tax_ids_exist = \"\"\"\n",
    "    MATCH (a:Taxon)\n",
    "    WHERE a.taxId in $tax_ids\n",
    "    RETURN COLLECT(DISTINCT a.taxId) as tax_ids\n",
    "\"\"\"\n",
    "\n",
    "query_palmprint_exists = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (a:SOTU)-[:HAS_POTENTIAL_TAXON]->(b:Taxon)\n",
    "        WHERE b.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as palmprints\n",
    "        UNION\n",
    "        MATCH (a:SOTU)<-[:HAS_SOTU*]-(b:Palmprint)-[:HAS_POTENTIAL_TAXON]->(c:Taxon)\n",
    "        WHERE c.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) + COLLECT(DISTINCT b.palmId) as palmprints\n",
    "    }\n",
    "    WITH tax_id, palmprints\n",
    "    RETURN\n",
    "        tax_id,\n",
    "        palmprints\n",
    "\"\"\"\n",
    "\n",
    "query_sotus_exist_parent_child = \"\"\"\n",
    "    CALL {\n",
    "        MATCH (a:SOTU)-[:HAS_POTENTIAL_TAXON]->(b:Taxon)\n",
    "        WHERE b.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as sotus\n",
    "        UNION\n",
    "        MATCH (a:Taxon)<-[:HAS_PARENT*]-(b:Taxon)<-[:HAS_POTENTIAL_TAXON]-(c:SOTU)\n",
    "        WHERE a.taxId in $tax_ids\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as sotus\n",
    "        UNION\n",
    "        MATCH (a:Taxon)-[:HAS_PARENT*]->(b:Taxon)<-[:HAS_POTENTIAL_TAXON]-(c:SOTU)\n",
    "        WHERE a.taxId in $tax_ids\n",
    "        AND b.rank = 'species'\n",
    "        RETURN b.taxId as tax_id, COLLECT(DISTINCT a.palmId) as sotus\n",
    "    }\n",
    "    WITH tax_id, sotus\n",
    "    RETURN\n",
    "        tax_id,\n",
    "        sotus\n",
    "\"\"\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Neo4j dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_ids = list(merged_dropna.TaxID.unique())\n",
    "zoonotic_tax_ids = list(merged_dropna[merged_dropna['IsZoonotic'] == True].TaxID.unique())\n",
    "host_class_tax_ids = [\n",
    "    '40674', # [Mammal] Mammalia\n",
    "    '33090', # [Plant], Viridiplantae\n",
    "    '4751', # [Fungi]\n",
    "    '33630', # [Parasite], Alveolata\n",
    "    '2', # [Bacteria]\n",
    "    '10239', # [Virus], Viruses\n",
    "    '6656', # [Anthropod], Arthropoda\n",
    "    '7742' # [Vertabrata], Fish\n",
    "    '12908', # [Metagenome], Unclassified / Metagenome\n",
    "    # '1476529', # [Fish], Cyclostomata \n",
    "    # '7777' # [Fish], Chondrichthyes\n",
    "    # '1476750', # [Fish], fish environmental sample\n",
    "]\n",
    "\n",
    "def _log_df(df):\n",
    "    namespace = globals()\n",
    "    var_name = [name for name in namespace if namespace[name] is df]\n",
    "    print(var_name, df.shape)\n",
    "    print(df.head())\n",
    "\n",
    "\n",
    "def fetch_cached_df(query, params, filename, use_cache=True, log=False):\n",
    "\n",
    "    if os.path.exists(neo4j_data_path + filename) and use_cache:\n",
    "        df = pd.read_csv(neo4j_data_path + filename)\n",
    "        df = utils.deserialize_df(df)\n",
    "    else:\n",
    "        df = gds.run_cypher(query, params=params)\n",
    "        df.to_csv(neo4j_data_path + filename, index=False)\n",
    "    if log:\n",
    "        _log_df(df)\n",
    "    return df\n",
    "\n",
    "\n",
    "def get_neo4j_data(stat_threshold=0.8):\n",
    "    sotu_nodes = fetch_cached_df(\n",
    "        query_sotu_nodes,\n",
    "        {\n",
    "            'tax_ids': tax_ids,\n",
    "            'zoonotic_tax_ids': zoonotic_tax_ids,\n",
    "        },\n",
    "        f\"{stat_threshold}/sotu_nodes.csv\"\n",
    "    )\n",
    "    sotus = sotu_nodes['palmId'].unique().tolist()\n",
    "    sotu_msa_edges = fetch_cached_df(\n",
    "        query_sotu_msa_edges,\n",
    "        {\n",
    "            'tax_ids': tax_ids,\n",
    "            'sotus': sotus,\n",
    "        },\n",
    "        f\"{stat_threshold}/sotu_msa_edges.csv\"\n",
    "    )\n",
    "    sotu_nodes['taxId'] = sotu_nodes['taxId'].astype(int)\n",
    "\n",
    "    taxon_order_nodes = fetch_cached_df(\n",
    "        query_stat_taxon_order_nodes,\n",
    "        {\n",
    "            'sotus': sotus,\n",
    "            'host_class_tax_ids': host_class_tax_ids,\n",
    "            'stat_threshold': stat_threshold,\n",
    "        },\n",
    "        f\"{stat_threshold}/taxon_order_nodes.csv\",\n",
    "    )\n",
    "    taxon_order_nodes['taxId'] = taxon_order_nodes['taxId'].astype(int)\n",
    "    taxon_order_nodes['appId'] = taxon_order_nodes['appId'].astype(int)\n",
    "    # taxon_order_nodes = taxon_order_nodes.drop_duplicates(subset=['taxId'], keep='last')\n",
    "\n",
    "    has_host_order_stat_edges = fetch_cached_df(\n",
    "        query_has_host_order_stat,\n",
    "        {\n",
    "            'sotus': sotus,\n",
    "            'stat_threshold': stat_threshold,\n",
    "        },\n",
    "        f\"{stat_threshold}/has_host_order_stat_edges.csv\",\n",
    "    )\n",
    "    has_host_order_stat_edges['targetAppId'] = has_host_order_stat_edges['targetAppId'].astype(int)\n",
    "\n",
    "    has_host_class_label_map = fetch_cached_df(\n",
    "        query_host_class_label_map,\n",
    "        {\n",
    "            'host_tax_ids': has_host_order_stat_edges['targetAppId'].astype(str).unique().tolist(),\n",
    "            'host_class_tax_ids': host_class_tax_ids,\n",
    "            'stat_threshold': stat_threshold,\n",
    "        },\n",
    "        f\"{stat_threshold}/has_host_class_label_map.csv\",\n",
    "    )\n",
    "    print(stat_threshold)\n",
    "    print(\"sotu_nodes: \", neo4j_data['sotu_nodes'].shape)\n",
    "    print(\"sotu_msa_edges: \", neo4j_data['sotu_msa_edges'].shape)\n",
    "    print(\"taxon_order_nodes: \", neo4j_data['taxon_order_nodes'].shape)\n",
    "    print(\"has_host_order_stat_edges: \", neo4j_data['has_host_order_stat_edges'].shape)\n",
    "    print()\n",
    "\n",
    "    return {\n",
    "        'sotu_nodes': sotu_nodes,\n",
    "        'sotu_msa_edges': sotu_msa_edges,\n",
    "        'taxon_order_nodes': taxon_order_nodes,\n",
    "        'has_host_order_stat_edges': has_host_order_stat_edges,\n",
    "        'has_host_class_label_map': has_host_class_label_map,\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log_graph_shape(neo4j_data):\n",
    "    print(neo4j_data['has_host_order_stat_edges']['targetAppId'].unique().shape)\n",
    "    print(neo4j_data['has_host_class_label_map']['taxId'].unique().shape)\n",
    "    print(neo4j_data['has_host_class_label_map']['hostClass'].value_counts())\n",
    "\n",
    "\n",
    "def log_missing_nodes(neo4j_data):\n",
    "    # TODO: look into why some virus taxons in dataset have no matching SOTUs\n",
    "    # ex: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=11856&lvl=3&lin=f&keep=1&srchmode=1&unlock\n",
    "    # exists in genbank, check for any SOTUs mapped to genbank id?\n",
    "    missing_tax_ids = set(tax_ids) - set(neo4j_data['sotu_nodes']['taxId'].astype(str))\n",
    "    print(f\"Total virus tax_ids in dataset: {len(set(tax_ids))}\")\n",
    "    print(f\"Total SOTUs matching virus tax_ids in neo4j: {len(set(neo4j_data['sotu_nodes']['taxId']))}\")\n",
    "    print(f\"SOTUs missing virus tax_ids in neo4j: {len(missing_tax_ids)}\")\n",
    "    print()\n",
    "\n",
    "    # TODO: look into why some SOTUs have no host order stat edges\n",
    "    # Verify all/most are metagenomic, still, they should not be included as SOTUs \n",
    "    print(f\"Total SOTUs: {len(neo4j_data['sotu_nodes']['appId'].unique())}\")\n",
    "    print(f\"SOTUs with host order stat edges: {len(neo4j_data['has_host_order_stat_edges']['sourceAppId'].unique())}\")\n",
    "    sotus_missing_host = neo4j_data['sotu_nodes'].loc[~neo4j_data['sotu_nodes']['appId'].isin(neo4j_data['has_host_order_stat_edges']['sourceAppId'])]['appId'].unique()\n",
    "    print(f\"SOTUs missing host order stat edges: {len(sotus_missing_host)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_thresholds = [0.2, 0.5, 0.8]\n",
    "\n",
    "for val in stat_thresholds:\n",
    "    neo4j_data = get_neo4j_data(stat_threshold=val)\n",
    "    log_graph_shape(neo4j_data)\n",
    "    log_missing_nodes(neo4j_data)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distribution investigation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Relationship between STAT host associations and IsZoonotic label\n",
    "\n",
    "Host associations are taken from [STAT](https://www.ncbi.nlm.nih.gov/sra/docs/sra-taxonomy-analysis-tool/) data from the SRA run. I pooled in the STAT host associations from all SRA runs associated to all palmprints associated to SOTUs in the dataset. The overlap between these SOTUs with STAT associations to primates (order-level homo sapiens) that also have a true IsZoonotic label using the zoonotic dataset has ~0.35 Jaccard similarity. \n",
    "\n",
    "This looks fairly low, but I'd need to do a deeper analysis to understand the significance. It would also make sense to incorporate the taxonomic/phylogenetic distance instead of using binary similary test.\n",
    "\n",
    "Overall, I'll be exploring the potential for significant signals in the host distribution, tissue distribution, and other network topology metrics that could be specific to zoonotic potential (next thread). I haven't found any research on the relationship between STAT host association and pathogenesis.\n",
    "\n",
    "\n",
    "```md\n",
    "A = has_primate_assoc_tax\n",
    "B = isZoonotic\n",
    "\n",
    "Using Virus taxons: \n",
    "| B\\A | 0    | 1    |\n",
    "| 0   | 0.36 | 0.23 | \n",
    "| 1   | 0.38 | 0.35 |\n",
    "\n",
    "Using SOTUs:\n",
    "| B\\A | 0    | 1    |\n",
    "| 0   | 0.28 | 0.17 | \n",
    "| 1   | 0.31 | 0.14 |\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jaccard similarity: 0.43529411764705883,  0.2535211267605634, 0.4370860927152318, 0.336283185840708\n",
      "unique SOTU taxIds with primate assoc 74\n",
      "unique SOTU taxIds with isZoonotic 77\n",
      "SOTUs taxIds in both 38\n",
      "SOTUs taxIds unique to isZoonotic 39\n",
      "SOTUs taxIds unique to primate assoc 36\n",
      "\n",
      "Jaccard similarity: 0.3041958041958042,  0.17765042979942694, 0.4715821812596006, 0.15584415584415584\n",
      "unique SOTUs with primate assoc 146\n",
      "unique SOTUs with isZoonotic 477\n",
      "SOTUs in both 84\n",
      "SOTUs unique to isZoonotic 393\n",
      "SOTUs unique to primate assoc 62\n"
     ]
    }
   ],
   "source": [
    "# Investigate difference between dataset isZoonotic label and human/primate STAT associations in SRA\n",
    "# TODO: create proper figure for this \n",
    "# TODO: use all pathogen data from biotic interaction db and find jaccard similarity, already has ncbi tax_id mapping\n",
    "\n",
    "# SOTUs have majority class: only 181 unique SOTU taxIds, taxId 145856 has 215 of 477 SOTUs\n",
    "# Seems better to use virus taxon\n",
    "# unique_sotu_tax_orders =  neo4j_data['sotu_nodes']['taxId'].value_counts()\n",
    "# print(unique_sotu_tax_orders)\n",
    "\n",
    "def jaccard_similarity(list1, list2):\n",
    "    s1 = set(list1)\n",
    "    s2 = set(list2)\n",
    "    return float(len(s1.intersection(s2)) / len(s1.union(s2)))\n",
    "\n",
    "has_primate_assoc = neo4j_data['has_host_order_stat_edges'].loc[neo4j_data['has_host_order_stat_edges']['targetAppId'] == 9443]['sourceAppId'].unique()\n",
    "has_primate_assoc_tax = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['appId'].isin(has_primate_assoc)]['taxId'].unique()\n",
    "has_zoonotic_potential = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['isZoonotic'] == True]['appId'].unique()\n",
    "has_zoonotic_potential_tax = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['appId'].isin(has_zoonotic_potential)]['taxId'].unique()\n",
    "\n",
    "not_has_primate_assoc = neo4j_data['has_host_order_stat_edges'].loc[neo4j_data['has_host_order_stat_edges']['targetAppId'] != 9443]['sourceAppId'].unique()\n",
    "not_has_primate_assoc_tax = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['appId'].isin(not_has_primate_assoc)]['taxId'].unique()\n",
    "not_has_zoonotic_potential = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['isZoonotic'] == False]['appId'].unique()\n",
    "not_has_zoonotic_potential_tax = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['appId'].isin(not_has_zoonotic_potential)]['taxId'].unique()\n",
    "\n",
    "tl = jaccard_similarity(not_has_primate_assoc_tax, not_has_zoonotic_potential_tax)\n",
    "tr = jaccard_similarity(has_primate_assoc_tax, not_has_zoonotic_potential_tax)\n",
    "bl = jaccard_similarity(not_has_primate_assoc_tax, has_zoonotic_potential_tax)\n",
    "br = jaccard_similarity(has_primate_assoc_tax, has_zoonotic_potential_tax)\n",
    "\n",
    "print(f'Jaccard similarity: {tl},  {tr}, {bl}, {br}')\n",
    "print(f'unique SOTU taxIds with primate assoc {len(has_primate_assoc_tax)}')\n",
    "print(f'unique SOTU taxIds with isZoonotic {len(has_zoonotic_potential_tax)}')\n",
    "print(f'SOTUs taxIds in both {len(set(has_zoonotic_potential_tax) & set(has_primate_assoc_tax)) }')\n",
    "print(f'SOTUs taxIds unique to isZoonotic {len(set(has_zoonotic_potential_tax) - set(has_primate_assoc_tax)) }')\n",
    "print(f'SOTUs taxIds unique to primate assoc {len(set(has_primate_assoc_tax) - set(has_zoonotic_potential_tax)) }')\n",
    "print()\n",
    "\n",
    "tl = jaccard_similarity(not_has_primate_assoc, not_has_zoonotic_potential)\n",
    "tr = jaccard_similarity(has_primate_assoc, not_has_zoonotic_potential)\n",
    "bl = jaccard_similarity(not_has_primate_assoc, has_zoonotic_potential)\n",
    "br = jaccard_similarity(has_primate_assoc, has_zoonotic_potential)\n",
    "print(f'Jaccard similarity: {tl},  {tr}, {bl}, {br}')\n",
    "print(f'unique SOTUs with primate assoc {len(has_primate_assoc)}')\n",
    "print(f'unique SOTUs with isZoonotic {len(has_zoonotic_potential)}')\n",
    "print(f'SOTUs in both {len(set(has_zoonotic_potential) & set(has_primate_assoc)) }')\n",
    "print(f'SOTUs unique to isZoonotic {len(set(has_zoonotic_potential) - set(has_primate_assoc)) }')\n",
    "print(f'SOTUs unique to primate assoc {len(set(has_primate_assoc) - set(has_zoonotic_potential)) }')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Zoonotic set enrichment analysis"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The main idea is to find significant correlations of traits (or other metrics) to zoonotic potential, starting with host range. To simplify the analysis, I'm using classes of hosts instead of specific tax_ids (Bacteria, Mammal, Virus, Plant, Fish, Anthropod, Fungi, Parasite, Other).\n",
    "\n",
    "The statistical tests will be similar to gene set enrinchment analysis (GSEA). Instead of genes, the host classes will be considered as classes in a functional class scoring (FCS) significance test. First I'll need to find the log2 fold changes between positive and negative groups, then I'll need to compute the enrichment score with significance values. It's not clear how well this will work for small sets, i'm also not sure if I can use existing GSEA libraries directly (DESeq2).\n",
    "\n",
    "Along with this test on hosts, I'm hoping to include topological metrics like modularity, triangle count, graph spectrum, graph magnitude,  to see if there are topological signatures related to zoonotic potential. The graph of a given SOTU would include: the SOTU, highly similar other SOTUs, the full taxonomy heirachy (?), relationships between the SOTU and the taxonomy heirarchy (STAT, virus tax id). We should be able to measure significance between the positive and negative groups using Fisher's test on these metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [],
   "source": [
    "# WANT: for all zoonotic SOTUs, get value_counts of hostClasses\n",
    "# Each row should be an SOTU, col for isZoonotic, col for each hostClass, count of the number of associations, avg > 0.2\n",
    "# Each row should be an SOTU, col for isZoonotic, col for each hostClass, avgPercentIdentity of the associations\n",
    "\n",
    "def generate_count_matricies(neo4j_data, threshold):\n",
    "    positive_sotu_nodes = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['isZoonotic'] == True]['appId'].unique()\n",
    "    negative_sotu_nodes = neo4j_data['sotu_nodes'].loc[neo4j_data['sotu_nodes']['isZoonotic'] == False]['appId'].unique()\n",
    "    total_sotu_nodes = neo4j_data['sotu_nodes']['appId'].unique()\n",
    "\n",
    "    target_nodes = [\n",
    "        positive_sotu_nodes,\n",
    "        negative_sotu_nodes,\n",
    "        total_sotu_nodes,\n",
    "    ]\n",
    "    target_names = [\n",
    "        'positive',\n",
    "        'negative',\n",
    "        'total',\n",
    "    ]\n",
    "\n",
    "    count_matricies = []\n",
    "\n",
    "    for target, name in zip(target_nodes, target_names):\n",
    "        has_host_df = neo4j_data['has_host_order_stat_edges'].loc[neo4j_data['has_host_order_stat_edges']['sourceAppId'].isin(target)]\n",
    "        taxon_nodes_df = neo4j_data['taxon_order_nodes'].loc[neo4j_data['taxon_order_nodes']['appId'].isin(has_host_df['targetAppId'])] \n",
    "        host_assoc_full_df =  has_host_df.merge(taxon_nodes_df, left_on='targetAppId', right_on='appId')\n",
    "        count_matrix = host_assoc_full_df.pivot_table(index='sourceAppId', columns='hostClass', values='count', aggfunc='sum', fill_value=0)\n",
    "        count_matrix.sort_index(inplace=True)\n",
    "        count_matrix.to_csv(neo4j_data_path + f\"{threshold}/\" + f\"count_matrix_{name}.csv\")\n",
    "\n",
    "        sotu_nodes_with_hosts = neo4j_data['sotu_nodes'].loc[\n",
    "            neo4j_data['sotu_nodes']['appId'].isin(has_host_df['sourceAppId'])\n",
    "        ]\n",
    "        metadata = pd.DataFrame({ \n",
    "            'sample': sotu_nodes_with_hosts['appId'].unique(),\n",
    "            'condition': sotu_nodes_with_hosts['isZoonotic'].astype(str).tolist(),\n",
    "        })\n",
    "        metadata.set_index('sample', inplace=True)\n",
    "        metadata.sort_index(inplace=True)\n",
    "        metadata.to_csv(neo4j_data_path +  f\"{threshold}/\" + f'count_metadata_{name}.csv')\n",
    "\n",
    "        for key in ['hostClass', 'taxOrder']:\n",
    "            value_counts = pd.concat([\n",
    "                host_assoc_full_df[key].value_counts(),\n",
    "                host_assoc_full_df[key].value_counts(normalize=True)],\n",
    "                axis=1,\n",
    "                keys=('count', 'percent'),\n",
    "            )\n",
    "            value_counts.to_csv(neo4j_data_path +  f\"{threshold}/\" + f\"{key}_value_counts.csv\")\n",
    "        count_matricies.append(count_matrix)\n",
    "    return count_matricies\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/owkin/PyDESeq2/blob/ae2d6b7efe77fcf80788ca62a9f845826c8b9a4d/pydeseq2/utils.py#L1420\n",
    "# Modifications to add labels to points\n",
    "\n",
    "def make_MA_plot(\n",
    "    results_df,\n",
    "    padj_thresh = 0.05,\n",
    "    log = True,\n",
    "    save_path = None,\n",
    "    lfc_null = 0,\n",
    "    alt_hypothesis = None,\n",
    "    **kwargs,\n",
    "):\n",
    "    colors = results_df[\"padj\"].apply(lambda x: \"darkred\" if x < padj_thresh else \"gray\")\n",
    "\n",
    "    fig, ax = plt.subplots(dpi=600)\n",
    "\n",
    "    # Set default alpha and s parameters, if not already specified\n",
    "    kwargs.setdefault(\"alpha\", 0.5)\n",
    "    kwargs.setdefault(\"s\", 0.2)\n",
    "\n",
    "    plt.scatter(\n",
    "        x=results_df[\"baseMean\"],\n",
    "        y=results_df[\"log2FoldChange\"],\n",
    "        c=colors,\n",
    "        **kwargs,\n",
    "    )\n",
    "\n",
    "    ax.set_adjustable(\"datalim\")\n",
    "\n",
    "    if log is True:\n",
    "        plt.xscale(\"log\")\n",
    "\n",
    "    plt.xlabel(\"mean of normalized counts\")\n",
    "    plt.ylabel(\"log2 fold change\")\n",
    "\n",
    "    plt.axhline(lfc_null, color=\"red\", alpha=0.5, linestyle=\"--\", zorder=3)\n",
    "    if alt_hypothesis and alt_hypothesis in [\"greaterAbs\", \"lessAbs\"]:\n",
    "        plt.axhline(-lfc_null, color=\"red\", alpha=0.5, linestyle=\"--\", zorder=3)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    offset = 0\n",
    "\n",
    "    for i in range(0, len(results_df[\"baseMean\"])):\n",
    "        plt.text(results_df[\"baseMean\"][i] - offset, results_df[\"log2FoldChange\"][i], f'{results_df.index[i]}')\n",
    "    \n",
    "\n",
    "    if save_path is not None:\n",
    "        plt.savefig(save_path, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_dseq_output(threshold):\n",
    "    full_count_matrix = pd.read_csv(neo4j_data_path + f\"{threshold}/\" + f\"count_matrix_total.csv\", index_col=0)\n",
    "    full_metadata = pd.read_csv(neo4j_data_path + f\"{threshold}/\" + f\"count_metadata_total.csv\", index_col=0)\n",
    "    \n",
    "    dds = DeseqDataSet(\n",
    "        counts=full_count_matrix,\n",
    "        metadata=full_metadata,\n",
    "        design_factors=\"condition\",\n",
    "        refit_cooks=True,\n",
    "    )\n",
    "    dds.deseq2()\n",
    "    stat_res = DeseqStats(dds)\n",
    "    stat_res.results_df.to_csv(neo4j_data_path + f\"{threshold}/\" + f\"deseq2_stats_res_no_lfc_shrink.csv\")\n",
    "    make_MA_plot(stat_res.results_df, s=20, save_path=neo4j_data_path + f\"{threshold}/\" + f\"deseq2_MA_plot_no_lfc_shrink.png\")\n",
    "    stat_res.lfc_shrink(coeff=\"condition_True_vs_False\")\n",
    "    stat_res.results_df.to_csv(neo4j_data_path + f\"{threshold}/\" + f\"deseq2_stats_res.csv\")\n",
    "    make_MA_plot(stat_res.results_df, s=20, save_path=neo4j_data_path + f\"{threshold}/\" + f\"deseq2_MA_plot.png\")\n",
    "    # print(dds)\n",
    "    # print(dds.varm[\"dispersions\"])\n",
    "    # print(dds.varm[\"LFC\"])\n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.2\n",
      "sotu_nodes:  (742, 8)\n",
      "sotu_msa_edges:  (2714, 7)\n",
      "taxon_order_nodes:  (281, 7)\n",
      "has_host_order_stat_edges:  (2950, 10)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "/Users/lukepereira/anaconda3/envs/rnalab/lib/python3.11/site-packages/pydeseq2/dds.py:356: RuntimeWarning: Every gene contains at least one zero, cannot compute log geometric means. Switching to iterative mode.\n",
      "  self.fit_size_factors()\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for threshold in stat_thresholds:\n",
    "    neo4j_data = get_neo4j_data(stat_threshold=threshold)\n",
    "    generate_count_matricies(neo4j_data, threshold)\n",
    "    generate_dseq_output(threshold)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Scatter bar plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_scatter_bar_plot(y, point_labels, ax, name=''):\n",
    "    np.random.seed(123)\n",
    "\n",
    "    w = 0.8\n",
    "    x = [1, 2]\n",
    "    colors = [(0, 0, 1, 1), (1, 0, 0, 1)] # Blue, Red\n",
    "\n",
    "    ax.set_title(name)\n",
    "\n",
    "    ax.set_yscale('log')\n",
    "    y_errs = [np.std(yi) for yi in y]\n",
    "    ax.bar(x,\n",
    "        height=[np.mean(yi) for yi in y],\n",
    "        yerr=y_errs, # error bars\n",
    "        capsize=12, # error bar cap width in points\n",
    "        width=w, \n",
    "        tick_label=[\"positive\", \"negative\"],\n",
    "        color=(0,0,0,0),  # face color transparent\n",
    "        edgecolor=colors,\n",
    "    )\n",
    "\n",
    "    for i in range(len(x)):\n",
    "        # distribute scatter randomly across whole width of bar\n",
    "        y_err = y_errs[i]\n",
    "        x_vals = x[i] + np.random.random(y[i].size) * w - w / 2\n",
    "        y_vals = y[i]\n",
    "        labels = point_labels[i]\n",
    "        ax.scatter(x_vals, y_vals, color=colors[i], s=10)\n",
    "        \n",
    "        for j in range(len(y_vals)):\n",
    "            y_val = y_vals[j]\n",
    "            x_val = x_vals[j]\n",
    "            label = labels[j]\n",
    "            if y_val > y_err:\n",
    "                ax.text(x_val, y_val -10, label, fontsize=8)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_count_matrix = pd.read_csv(neo4j_data_path + f\"{threshold}/\" + f\"count_matrix_total.csv\", index_col=0)\n",
    "random_count_matrix = total_count_matrix.sample(n=100, random_state=1)\n",
    "print(random_count_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAicAAAGzCAYAAAD0T7cVAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAABq6klEQVR4nO3deVxU5f4H8M+AgjsumKKCVoqBC6gJueUSZZqa2uJWoZktFzXjV5LXm0tpqN26Wk7ZdbfSLMt2tww1UyaUUMsKLVFyQTEdBFEUnt8fcxkZmOWcmTMz58x83q/XvGDOnDnnOc8588x3zrPphBACRERERCoR4O0EEBEREVXE4ISIiIhUhcEJERERqQqDEyIiIlIVBidERESkKgxOiIiISFUYnBAREZGqMDghIiIiVWFwQkRERKrC4IT80o4dO6DT6bBjxw5vJ4WIvCQnJwc6nQ6rVq3ydlKoEgYnfmbVqlXQ6XTQ6XTYvXt3ldeFEAgPD4dOp8OgQYO8kEIiUqOKZUf546abbkLfvn2xadMmt+337bffZvDgh6p5OwHkHTVq1MDatWvRs2dPi+U7d+7EX3/9heDgYC+ljIjU7OWXX8bNN98MIQTy8vKwatUqDBw4EF9++aVbftC8/fbbCA0NxdixYxXfdsuWLVFcXIzq1asrvm1yDYMTPzVw4EB8/PHHePPNN1Gt2o3LYO3atejSpQvy8/O9mDoiUqsBAwbg9ttvNz8fP348mjRpgnXr1mnmbuv169dRVlaGoKAg1KhRw9vJIStYreOnRo0ahfPnz2Pbtm3mZSUlJdiwYQNGjx5dZf1///vf6N69Oxo1aoSaNWuiS5cu2LBhQ5X1dDodJk6ciI8//hjR0dGoWbMmunXrhkOHDgEA3n33XbRu3Ro1atRAnz59kJOTY/H+Pn36oH379jh48CB69+6NWrVqoXXr1uZ97dy5E/Hx8ahZsybatm2Lb7/91uL9x48fxz/+8Q+0bdsWNWvWRKNGjfDQQw9V2Q8RKaN+/fqoWbOmxY8cqeUFALz//vuIi4tDrVq10KBBA9x5553YunUrAKBVq1b45ZdfsHPnTnNVUp8+fczvvXjxIqZMmYLw8HAEBwejdevWmD9/PsrKyszrlLcr+fe//42FCxfi1ltvRXBwMA4fPmy1zcnBgwcxduxY3HLLLahRowaaNm2Kxx9/HOfPn1c248gu3jnxU61atUK3bt2wbt06DBgwAACwadMmGI1GjBw5Em+++abF+osWLcKQIUMwZswYlJSU4MMPP8RDDz2Er776Cvfdd5/Fut9//z2++OILJCUlAQBSU1MxaNAgTJ06FW+//Tb+8Y9/4MKFC1iwYAEef/xxfPfddxbvv3DhAgYNGoSRI0fioYcewjvvvIORI0figw8+wJQpU/D0009j9OjReO211/Dggw8iNzcXdevWBQBkZGRgz549GDlyJFq0aIGcnBy888476NOnDw4fPoxatWq5K0uJ/ILRaER+fj6EEDh79izeeustFBYW4pFHHjGvI7W8mD17NmbNmoXu3bvj5ZdfRlBQEAwGA7777jvcc889WLhwISZNmoQ6depg+vTpAIAmTZoAAC5fvozevXvj5MmTeOqppxAREYE9e/Zg2rRpOH36NBYuXGiR7pUrV+LKlSt48sknERwcjIYNG1oEMeW2bduGP//8E+PGjUPTpk3xyy+/4L///S9++eUXpKenQ6fTuSFXqQpBfmXlypUCgMjIyBCLFy8WdevWFZcvXxZCCPHQQw+Jvn37CiGEaNmypbjvvvvM7ytfp1xJSYlo37696Nevn8VyACI4OFgcO3bMvOzdd98VAETTpk1FQUGBefm0adMEAIt1e/fuLQCItWvXmpf99ttvAoAICAgQ6enp5uVbtmwRAMTKlSttplMIIfbu3SsAiDVr1piXpaWlCQAiLS3NTm4RUbnysqPyIzg4WKxatcpiXSnlxZEjR0RAQIAYNmyYKC0ttVi/rKzM/H+7du1E7969q6TnlVdeEbVr1xbZ2dkWy1988UURGBgoTpw4IYQQ4tixYwKAqFevnjh79qzFuuWvOSpD1q1bJwCIXbt2WckZcgdW6/ixhx9+GMXFxfjqq69w6dIlfPXVV1ardACgZs2a5v8vXLgAo9GIXr16ITMzs8q6d911F1q1amV+Hh8fDwB44IEHzHc4Ki7/888/Ld5fp04djBw50vy8bdu2qF+/PqKioszvsfX+ium8du0azp8/j9atW6N+/fpW00pE8uj1emzbtg3btm3D+++/j759++KJJ57Ap59+al5HSnnx2WefoaysDDNmzEBAgOVXkZS7Ex9//DF69eqFBg0aID8/3/xISEhAaWkpdu3aZbH+Aw88gMaNGzvcbsW0X7lyBfn5+bjjjjsAgGWIB7Fax481btwYCQkJWLt2LS5fvozS0lI8+OCDVtf96quvMGfOHGRlZeHq1avm5dYKkYiICIvnISEhAIDw8HCryy9cuGCxvEWLFlW2GxISIun9xcXFSE1NxcqVK3Hy5EkIIcyvGY1Gq8dGRNLFxcVZNIgdNWoUOnXqhIkTJ2LQoEEICgqSVF788ccfCAgIQHR0tFPpOHLkCA4ePGgz4Dh79qzF85tvvlnSdv/++2/Mnj0bH374YZVtsAzxHAYnfm706NGYMGECzpw5gwEDBqB+/fpV1vn+++8xZMgQ3HnnnXj77bcRFhaG6tWrY+XKlVi7dm2V9QMDA63uy9byigGEq++fNGkSVq5ciSlTpqBbt24ICQmBTqfDyJEjrdYvE5FrAgIC0LdvXyxatAhHjhzB33//Lau8cFZZWRnuvvtuTJ061errkZGRFs8r3hGx5+GHH8aePXvwwgsvIDY2FnXq1EFZWRnuvfdeliEexODEzw0bNgxPPfUU0tPTsX79eqvrfPLJJ6hRowa2bNliMf7JypUrPZVMyTZs2IDExES8/vrr5mVXrlzBxYsXvZcoIh93/fp1AEBhYaHk8uLWW29FWVkZDh8+jNjYWJvbtlXFc+utt6KwsBAJCQmuH8D/XLhwAdu3b8fs2bMxY8YM8/IjR44otg+Shm1O/FydOnXwzjvvYNasWRg8eLDVdQIDA6HT6VBaWmpelpOTg88++8xDqZQuMDCwyp2Yt956yyLtRKSca9euYevWrQgKCkJUVJTk8mLo0KEICAjAyy+/XOWORMXPcO3ata3+uHj44Yexd+9ebNmypcprFy9eNAdMcpTfna1chlTu+UPuxzsnhMTERLuv33fffXjjjTdw7733YvTo0Th79iz0ej1at26NgwcPeiiV0gwaNAjvvfceQkJCEB0djb179+Lbb79Fo0aNvJ00Ip+wadMm/PbbbwBM7TrWrl2LI0eO4MUXX0S9evUklxetW7fG9OnT8corr6BXr14YPnw4goODkZGRgWbNmiE1NRUA0KVLF7zzzjuYM2cOWrdujZtuugn9+vXDCy+8gC+++AKDBg3C2LFj0aVLFxQVFeHQoUPYsGEDcnJyEBoaKuvY6tWrhzvvvBMLFizAtWvX0Lx5c2zduhXHjh1TLgNJEgYn5FC/fv2wfPlyzJs3D1OmTMHNN9+M+fPnIycnR3XByaJFixAYGIgPPvgAV65cQY8ePfDtt9+if//+3k4akU+oWN1Ro0YN3HbbbXjnnXfw1FNPAZBXXpQPhf/WW29h+vTpqFWrFjp27IhHH33UYn/Hjx/HggULcOnSJfTu3Rv9+vVDrVq1sHPnTrz66qv4+OOPsWbNGtSrVw+RkZGYPXu2ucG8XGvXrsWkSZOg1+shhMA999yDTZs2oVmzZk5tj5yjE5XvXxERERF5EducEBERkaowOCEiIiJVYXBCREREqsLghIiIiFSFwQkRERGpCoMTIiIiUhXNjXNSVlaGU6dOoW7dupJmriQi5QkhcOnSJTRr1qzKjLJqxbKDyLvklBuaC05OnTpVZXZaIvKO3NxctGjRwtvJsEuv10Ov16OkpAR//PGHt5ND5PeklBuaG4TNaDSifv36yM3NRb169bydHCK/VFBQgPDwcFy8eNHpkTg9jWUHkXfJKTc0d+ek/HZsvXr1WMAQeZmWqkdYdhCpg5RyQxuVxUREROQ3GJwQERGRqngtOLl8+TJatmyJ559/3ltJICIiIhXyWnAyd+5c3HHHHd7aPREREamUV4KTI0eO4LfffsOAAQO8sXsiIiJSMdnBya5duzB48GA0a9YMOp0On332WZV19Ho9WrVqhRo1aiA+Ph4//vijxevPP/88UlNTnU40ERER+S7ZwUlRURFiYmKg1+utvr5+/XokJydj5syZyMzMRExMDPr374+zZ88CAD7//HNERkYiMjLStZST08aNG4fIyEjExMSgR48eyMjIML92+fJljBo1Cq1bt0ZkZCQ2bNhgfm3YsGGIjY01PwICAvDFF18AMI2+OWnSJNx6661o3bo1Fi9eXGW/Z8+eRZMmTTB06NAqrxUXFyM6OhqxsbGKHy8RuY7lBnmUcAEAsXHjRotlcXFxIikpyfy8tLRUNGvWTKSmpgohhHjxxRdFixYtRMuWLUWjRo1EvXr1xOzZs23u48qVK8JoNJofubm5AoAwGo2uJN2vff755+LatWtCCCG+/PJL0bJlS/Nrs2fPFomJiUIIIf7880/RuHFjkZ+fX2UbGRkZolGjRuLq1atCCCFWr14t+vXrJ65fvy7Onz8vIiIixM8//2zxnqFDh4rHH39c3H///VW2N2nSJPHEE0+ImJgYRY6R3MtoNGruc6jFNKsJyw1ylZzPoKJtTkpKSrB//34kJCSYlwUEBCAhIQF79+4FAKSmpiI3Nxc5OTn497//jQkTJmDGjBk2t5mamoqQkBDzg0PXS6PT6XDx4kXz89DQUOTk5AAAhgwZgmrVTOPv3XHHHTh58iSuX78OwHTn6+mnnwYA3HzzzejTpw82btxYZfvLly/HI488gqCgIPP7JkyYgMDAQDRs2BAjRozAunXrLNa/+eab0atXryrb+vbbb3Hy5EmMGTNGkWMnIuew3CC1UDQ4yc/PR2lpKZo0aWKxvEmTJjhz5oxT25w2bRqMRqP5kZubq0RS6X8WLVqEgQMHmgudEydOoGXLlubXW7VqhRMnTli8p7i4GOvWrcP48ePNy+y979ixY1iyZAnmzp1bZf8XL17E1KlT8c477yh6XETkPiw3yN28Onz92LFjHa4THByM4OBg9yfGRxgMQHa2tHXff/99fPTRR9i1a5esfWzYsAGRkZHo0KGDw3WFEHj88cexePFi1KxZs8rrEydOxD//+U/cdNNNOHz4sKx0EJEyWG6Q2ih65yQ0NBSBgYHIy8uzWJ6Xl4emTZsquSuyIiUFuOMO4LHHACAQM2aUml+7cuWKxbrr16/H7NmzsW3bNos7XRERETh+/Lj5eU5ODiIiIizeu3z5cotfP/beV1BQgIMHD2LEiBFo1aoVnn/+eWzduhV33XUXAGD37t14/vnn0apVK4wcORKHDx9G27ZtXc0KIpKI5QapkiuNW2CjQezEiRPNz0tLS0Xz5s3NDWKdtXjxYhEVFSUiIyPZqM2K9HQhgIqPtgL4WqSnC/HJJ58IAOLYsWNCCCHWr18vWrduLXJycqpsZ+bMmVUatp07d878+pEjR0Tt2rWr5P/KlSurNGw7ePBgle2vXLnSasM2IYRIS0tjwzaN0GLjUi2m2d1YbpAnubVBbGFhIbKyspCVlQXAVC+YlZVlridMTk7G0qVLsXr1avz666945plnUFRUhHHjxrkURCUlJeHw4cMW3dfohqq3ZP8D4FmMGNEZP/30Exo1amR+ZcyYMbhy5Qruv/9+c/e+8+fPAwBeeOEFFBcX49Zbb0X//v2xePFihIaGmt+7YsUKPPDAA1VmdX300Udx2223oU2bNujatSuSk5Ml3b4lIu9huUFqpRNCCDlv2LFjB/r27VtleWJiIlatWgUAWLx4MV577TWcOXMGsbGxePPNNxEfH69IggsKChASEgKj0chpzyswGEy3ZitLTwcUynoiMy1+DrWYZndjuUGeJOczKPvOSZ8+fSCEqPIoD0wAU2Ol48eP4+rVqzAYDIoFJmRbfDwwdarlspQUFjBEer0e0dHR6Nq1q7eTojosN0itZN858Ra9Xg+9Xo/S0lJkZ2fz148N5a3uIyNZwJD7aPEuhBbT7CksN8gT5HwGNROclGMBQ+R9WvwcajHNRL7ErdU6RERERO7E4ISIiIhUhcEJERERqYpmghO2uCciIvIPmglOOAgbERGRf9BMcEJERET+gcEJERERqQqDEyIiIlIVBidERESkKpoJTthbh4iIyD9oJjhhbx0iIiL/oJnghIiIiPwDgxMiIiJSFQYnREREEsXHxyM2NhaxsbFo3749dDodDh48CAA4cuQI7r77bsTExKBdu3ZYv369+X3nzp3D4MGD0bFjR0RFRSExMRHFxcXm13fu3ImuXbuiXbt2iI6Oxt69ez1+bGrC4ISIiEgig8GArKwsZGVlYdasWWjfvj06duwIABg7dixGjBiBAwcOYMeOHZg6dSpOnjwJAJg7dy7atGmDgwcP4ueff0ZeXh5WrlwJADh16hQSExOxZs0a/PLLL/jpp58QFRXltWNUA80EJ+ytQ0REnqDT6XDx4kXz89DQUOTk5FRZb/ny5Rg/frz5+YEDBzBw4EAAQOPGjRETE2O+e6LT6XDp0iWUlZWhpKQEly9fRosWLQAAb7/9NkaPHm0OSIKDg1G/fn33HJxGaCY4YW8dIiJSi9zcXOzcuROPPPKIeVmXLl3w/vvvAwD+/PNP7NmzxxzUvPTSSzh69CiaNm2Km266CVFRURgyZAgA4PDhwyguLkZCQgJiY2MxadIkFBUVefyY1EQzwQkREZFbGQzAe+9JWnXVqlUYNGgQQkNDzctWr14Ng8GA2NhYPPfcc7jrrrtQrVo1AMCHH36I6OhonD59GqdOnUJ2djaWLVsGALh+/Tp27dqFjz/+GBkZGbhw4QJmzpyp/PFpCIMTIiKilBTgjjuAxx5DIIDSGTPML125csViVSEEVq5caVGlAwCtWrXCJ598gqysLHz++ecwGo1o164dAFPVzZgxYxAYGIi6deviwQcfRFpaGgAgIiIC9913Hxo0aIDq1atj1KhRSE9Pd+/xqhyDEyIi8m8GA7BggflpawCGt94CDAZ8+umnVapYvvvuO1y/fh133323xfK8vDyUlZUBALZs2YLDhw9j9OjRAIBbbrkFmzdvBgBcu3YNW7ZsQfv27QEAo0ePRlpaGq5evQoA2LRpE2JiYtxyqFrB4ISIiPxbdrbF0/8AeBZA5xEj8NNPP6FRo0YWry9fvhzjxo1DQIDlV+iXX36JyMhItG3bFvPmzcM333yDmjVrAgAWLVoEg8GADh06ICYmBo0bN8Zzzz0HAOjevTuGDBmCTp06oUOHDsjPz8fcuXPddrhaoBNCCG8nQo6CggKEhITAaDSiXr163k4OkV/S4udQi2kmDzEYTFU6laWnA/Hxnk+Pj5LzGeSdEyLShIsXL+L22283D361dOlSbyeJfEV8PDB1quWylBQGJl5UzdsJICKSom7duti1axdq1aqFoqIitG/fHsOHD69yy53IKfPnA8OHm6p4IiMZmHiZZoITvV4PvV6P0tJSbyeFiLwgMDAQtWrVAgBcvXoVQghorFaa1C4+nkGJSmimWoeDsBFp265duzB48GA0a9YMOp0On332WZV19Ho9WrVqhRo1aiA+Ph4//vijxesXL15ETEwMWrRogRdeeMFijAki8h2aCU6ISNuKiooQExMDvV5v9fX169cjOTkZM2fORGZmJmJiYtC/f3+cPXvWvE79+vVx4MABHDt2DGvXrkVeXp7N/V29ehUFBQUWDyLSBgYnROQRAwYMwJw5czBs2DCrr7/xxhuYMGECxo0bh+joaCxZsgS1atXCihUrqqzbpEkTxMTE4Pvvv7e5v9TUVISEhJgf4eHhih0LEbkXgxMi8rqSkhLs378fCQkJ5mUBAQFISEgwTx2fl5eHS5cuAQCMRiN27dqFtm3b2tzmtGnTYDQazY/c3Fz3HgQRKUYzDWKJyHfl5+ejtLQUTZo0sVjepEkT/PbbbwCA48eP48knnzQ3hJ00aRI6dOhgc5vBwcEIDg52a7qJyD0YnBCRJsTFxSErK8vbySAiD2C1DhF5XWhoKAIDA6s0cM3Ly0PTpk29lCoi8hYGJ0TkdUFBQejSpQu2b99uXlZWVobt27ejW7duLm1br9cjOjoaXbt2dTWZROQhrNYhIo8oLCzE0aNHzc+PHTuGrKwsNGzYEBEREUhOTkZiYiJuv/12xMXFYeHChSgqKsK4ceNc2m9SUhKSkpLM83oQkfppJjjhCLFE2rZv3z707dvX/Dw5ORkAkJiYiFWrVmHEiBE4d+4cZsyYgTNnziA2NhabN2+u0kiWiHwfZyUmItm0+DnUYpqJfAlnJSYiIiLNYnBCRD6NDWKJtIfBCRH5NE4aSqQ9DE6IiIhIVRicEBERkaowOCEiIiJVYXBCREREqsLghIh8GnvrEGkPgxMi8mnsrUOkPQxOiIiISFUYnJBfio+PR2xsLGJjY9G+fXvodDocPHgQAHDkyBHcfffdiImJQbt27bB+/Xrz+86dO4fBgwejY8eOiIqKQmJiIoqLi82v79y5E127dkW7du0QHR2NvXv3evzYiIi0jsEJ+SWDwYCsrCxkZWVh1qxZaN++PTp27AgAGDt2LEaMGIEDBw5gx44dmDp1Kk6ePAkAmDt3Ltq0aYODBw/i559/Rl5eHlauXAkAOHXqFBITE7FmzRr88ssv+OmnnxAVFeW1YyQi0ioGJ+SzdDodLl68aH4eGhqKnJycKustX74c48ePx6uvvoq2bdtiz549CAgwfTQaN26MmJgYvPbaa7jjjjuwdu1arFixAvPmzUNJSQkuX76MFi1aAABefvllnDx5Eu+++y4AIDg4GAsXLkTjxo3Nd2nGjBlj3m9ZWRkmTZqEW2+9Fa1bt8bixYvdlxlERBqimVmJ9Xo99Ho9SktLkZ2dzZlFySGdTocLFy6gfv36AEzByb59+9CqVSvzOrm5uWjbti1OnDiBP//8E6GhoejcuTPuv/9+rF69Gn/++Sfi4uKg0+mwfPly9OzZE/fffz9++OEH1KxZE4888gjeffddXLt2DWFhYQgNDcXly5fRsGFD9OrVC3Xr1sXly5excOHCKulbs2YNVq9eja1bt8JoNKJTp0745ptv0K5dO89kkAu0OMOvFtNM5Et8clZitrgnJdx8880Wd1Patm2Lvn37IjQ0FHFxcbjllltw2223ITs7G7GxsXjuuedw1113me/CfPjhh7jlllsQFhaGAwcOIDs7G8uWLcPLL7+MsLAwFBQUYODAgcjIyMCFCxewY8cOm2lZv349JkyYgMDAQDRs2BAjRozAunXr3J8JfoZdiYm0p5q3E0CkOIMByM5GYGAgSktLzYuvXLlisZoQAlevXsXDDz9ssbxGjRqYMmUKhg4dCgC499578cwzz+Cll17CqVOnEBAQgKVLl6J169Z48MEH8fHHH6O0tBR33nknfv75Z9SoUQPVq1fHqFGj8Mwzz+D48eNIS0tDo0aN8NJLL6Fv374AgBMnTqBly5bm/bZq1Qrp6eluyhT/lZSUhKSkJPOvNiJSPwYn5FtSUoAFCwAArQEYnnoKAzdswKfz5qGoqMhi1e+++w4A0KtXL4vlJSUlKCsrAwBs2bIFhw8fRp06dZCamooPP/wQLVu2xPTp0xETE4NvvvkGP//8M3bv3o3Tp09j1KhR6NChAwBg06ZNuOuuu/Df//4X1atXxw8//IBhw4YhIyPDIighIiJLmqnWIXLIYDAHJgDwHwDPfvIJOjdsiJ+mTUMjmC740hkzAJgawgYGBpobv5Y7f/48kpKS0LZtW8ybNw8ffPABvvzyS4wePRqLFi1CdnY2CgoKMGDAAAQHB+Pq1avo27cvRo8ejbNnz+Kdd95BgwYNkJ+fj//85z+oXr06AKBHjx7o1KkT9u3bBwCIiIjA8ePHzfvNyclBRESEW7OIiEgLGJyQ78jOtng6AMARAJkXLuAVAPkA2gAwvPUWYDDgwQcfxLVr16psJiwsDO+88w5+//13pKWloXv37qhduza+++473Hzzzfjggw8QEhKCTz75BJ9++iny8/ORk5ODnJwcpKSkYNKkSbhw4QI+/PBDFBYWmrd75MgRZGVlme+sPPTQQ1i6dClKS0vx999/Y/369RgxYoTbsoeISCtYrUO+IzLS4Sr/ATAZwL9GjMB9jz6KRo0amV+bM2cOlixZgnPnzuHnn3/GxIkT8dNPP6Fx48b46KOP8MILL+D69eu4du0apkyZgm7dujnc3/Tp07F//35Uq1YNgYGB0Ov1iPxfOh999FFkZGSgTZs20Ol0SE5ONgcuRET+TDNdicuxOyDZVaHNCQDg0UeB996rul56OhAf77l0+Rgtfg61mGYiXyLnM8g7J+Rb5s8Hhg83VfFERpoCkLAwy4AlJYWBCRGRijE4Id8TH28ZfFgLWMhvVBzAkYi0gdU6RCSbFj+HWkwzkS/xyRFiiYiIyD8wOCEiIiJVYXBCREREqsLghIiIiFSFwQkRERGpCoMTIiIiUhUGJ0RERKQqDE6IiIhIVRicEJFP0+v1iI6ORteuXb2dFCKSiMEJEfm0pKQkHD58GBkZGd5OChFJ5PHg5OLFi7j99tsRGxuL9u3bY+nSpZ5OAhEREamYxyf+q1u3Lnbt2oVatWqhqKgI7du3x/Dhw9GoUSNPJ4WIiIhUyON3TgIDA1GrVi0AwNWrVyGEgMbmHiQiIiI3kh2c7Nq1C4MHD0azZs2g0+nw2WefVVlHr9ejVatWqFGjBuLj4/Hjjz9avH7x4kXExMSgRYsWeOGFFxAaGur0ARAREZFvkR2cFBUVISYmBnq93urr69evR3JyMmbOnInMzEzExMSgf//+OHv2rHmd+vXr48CBAzh27BjWrl2LvLw8m/u7evUqCgoKLB5ERETku2QHJwMGDMCcOXMwbNgwq6+/8cYbmDBhAsaNG4fo6GgsWbIEtWrVwooVK6qs26RJE8TExOD777+3ub/U1FSEhISYH+Hh4XKTTERERBqiaJuTkpIS7N+/HwkJCTd2EBCAhIQE7N27FwCQl5eHS5cuAQCMRiN27dqFtm3b2tzmtGnTYDQazY/c3Fwlk0xEREQqo2hvnfz8fJSWlqJJkyYWy5s0aYLffvsNAHD8+HE8+eST5oawkyZNQocOHWxuMzg4GMHBwUomk4iIiFTM412J4+LikJWV5endEhERkUYoWq0TGhqKwMDAKg1c8/Ly0LRpU5e2zSGoiYiI/IOiwUlQUBC6dOmC7du3m5eVlZVh+/bt6Natm0vb5hDUROQM/rAh0h7Z1TqFhYU4evSo+fmxY8eQlZWFhg0bIiIiAsnJyUhMTMTtt9+OuLg4LFy4EEVFRRg3bpyiCScikiIpKQlJSUkoKChASEiIt5NDRBLIDk727duHvn37mp8nJycDABITE7Fq1SqMGDEC586dw4wZM3DmzBnExsZi8+bNVRrJEhEREVmjExoZO16v10Ov16O0tBTZ2dkwGo2oV6+et5NF5JfK70Jo6XOoxTQT+RI5n0GPz63jLLY5ISIi8g+aCU6IiIjIPzA4ISIiIlVhcOKijIwM9OjRAzExMYiNjcV3331nfu2f//wnbrvtNsTExOD222/Hli1bzK+dO3cOgwcPRseOHREVFYXExEQUFxdbbLu4uBjR0dGIjY01L8vJyUGfPn0QEhJisRwwddt+/vnn0b59e9x2220YP348SkpK3HLcRERE7qKZ4ESNYxUIITBs2DDMnj0bBw4cwEcffYSxY8eag4xevXrhp59+woEDB7B8+XI8/PDDKCoqAgDMnTsXbdq0wcGDB/Hzzz8jLy8PK1eutNh+SkoKevToYbGsXr16mDNnDtauXVslPcuXL0dmZiYyMzPx66+/IiAgAIsWLXLT0RMREbmHZoITbzaI1el0uHjxovl5aGgocnJycP78eZw7d8480WFkZCTq16+PTZs2ATDN4FyzZk0AQIcOHSCEwLlz58zbvHTpEsrKylBSUoLLly+jRYsW5n18++23OHnyJMaMGWORloYNG6Jnz56oXbt2lXQeOHAACQkJCAoKgk6nw4ABA/Dee+8pmhdERETuppngxBsMBsDed3toaCjCwsLw0UcfATBV8fz+++/Iycmpsu7KlStxyy23oGXLlgCAl156CUePHkXTpk1x0003ISoqCkOGDAEAXLx4EVOnTsU777wjK71dunTBF198gYKCAly7dg0fffSR1bQQERGpGYMTG1JSgDvuAB57zPR85kzr633++edYsWIFOnXqhEWLFqFnz56oVs1ybLvt27dj9uzZWL9+PXQ6HQDgww8/RHR0NE6fPo1Tp04hOzsby5YtAwBMnDgR//znP3HTTTfJSvPYsWNx7733onfv3ujduzciIyOrpIWIiEjt+M1lhcEALFhQcUkg3nyzFKNHA/HxwJUrV8yvxMTEYPPmzebnbdq0QVBQEDIzMwEA+/fvx0svvWQexr98+RtvvIHp06fjwIEDAID4+Hh88skn6Ny5M7777jt89913mDx5MkpKSmA0GtGqVSt8+umn5v1kZ2ejuLjYvL1yQ4YMwZAhQxAWFoadO3eiXbt2CucOERGRezE4sSI7u/KS1gAMyM4eiJMnPzU3agWA06dPIywsDACwdOlSFBUV4ZlnnqmyzVGjRlVZ9vjjj1dZVjHQqej48ePo0qVLleXWlgHACy+8gK1bt+KVV16x+joREZFaaSY4qTh8vbtFRlZe8h8AkzFnzr/w8MP3oVGjRuZX/vvf/+KDDz6AEAJRUVH44osvEBBgqi0bOnQoioqKEBoaal7/5ZdfRps2bXDy5Em8+uqrOHfuHMrKytCxY0dMnjwZd911FwBg9+7dqFmzJvbt24fXX38d69atA2DqXjx8+HCUlJSgsLAQDRs2xMCBAzFp0iScP38eTz75JAICArBx40b83//9HwYPHuzOrCIiIlKcZubWKeep+TFSUiyrdlJSgHnz3LY7AEBRURHq1KkDwDT7s7UeOURqoMV5arSYZiJfIuczqJk7J542fz4wfLipiicy0tTWhIiIiNyPwYkd8fEMSoiIiDyNXYmJyKepcXRpIrKPwQkR+TRvji5NRM5hcEJERESqopnghLdmiYiI/INmghPemiUiIvIPmglOiIiIyD8wOCEiIiJVYXBCREREqsLghIiIiFSFwQkRERGpCoMTIiIiUhXNBCcc54SIiMg/aCY44TgnRERE/oGzEnuIwQBkZwORkZzpmIgsnT59GqdPn/Z2MhAWFoawsDBvJ4OIwYknpKQACxbceD51KjB/vvfSQ0Tq8u6772L27NneTgZmzpyJWbNmeTsZRAxO3M1gsAxMANPz4cN5B4WITJ566ikMGTLEqfcWFxejZ8+eAIDdu3ejZs2aTqeDd01ILRicuFl2tu3lDE6ICHCtOqWoqMj8f2xsLGrXrq1Usoi8RjMNYrUqMlLeciIiIn/H4MTN4uNNbUwqSknhXRMiIiJbWK3jAfPnm9qYsLcOERGRYwxOPCQ+nkEJERGRFKzWISIiIlXRTHDC4euJiIj8g2aCEw5fT0RE5B80E5wQeUJGRgZ69OiBmJgYxMbG4rvvvjO/duTIEdx9992IiYlBu3btsH79evNrs2bNQuPGjREbG4vY2FiMGTPG/Jper0eHDh0QGxuL9u3b48033zS/9vXXX6NLly4IDg7GlClTPHKMcr366qto27YtAgIC8Nlnn1m8JoTArFmzEBkZiQ4dOqBv377m13788Ufccccd6NSpE6KiorCgwmiEly9fxqhRo9C6dWtERkZiw4YNnjocItICoTFGo1EAEEaj0dtJUVxhYaEAIACIwsJCbyfH75SVlYnmzZuLbdu2CSGE+P3330V4eLi4fPmyEEKI7t27i6VLlwohhDh79qyIiIgQf/31lxBCiJkzZ4pnn33W6nYvXrxo/t9oNIrw8HCRmZlp3kdWVpaYPn26zfd7m8FgEH/88Yfo3bu32LhxoxDixudw3rx5YtiwYeLq1atCCCFOnz5tfl9MTIz4/PPPhRBCnD9/XjRu3Fj88ssvQgghZs+eLRITE4UQQvz555+icePGIj8/3246Tpw4IXr37i2ioqJEhw4dxEcffSTrOHy17GC5QVoh5zPIOyfkd3Q6HS5evGh+HhoaipycHJw/fx7nzp1DQkICACAyMhL169fHpk2bAAAHDhzAwIEDAQCNGzdGTEyMxd0TW0JCQsz/FxUV4dq1a+bnkZGRiImJQbVq3u04ZytPACAuLg633HKL1fe9+eabmDdvHoKCggAATZs2tbrNoqIiBAUFoWHDhgCA9evX4+mnnwYA3HzzzejTpw82btxoN43VqlXDwoULcfjwYWzduhVTpkyxGB2ViHwHgxPyHwYD8N57Nl8ODQ1FWFgYPvroIwCmKp7ff//d/CXdpUsXvP/++wCAP//8E3v27DG/BgAff/wxYmJi0K9fP6SlpVlse8OGDWjXrh1atWqF559/Hp06dVL22JzlIE8cOXv2LD7//HPEx8cjPj7eIlhbuXIlXnrpJURERCAyMhKvvvqqOXg5ceIEWrZsaV63VatWOHHihN19hYWFITY2FoApCAoNDcXff//tdNqJSL04zgn5h8pTQ8+cCSxaVGW1zz//HCkpKUhNTUW7du3Qs2dP812N1atX4//+7/8QGxuLli1b4q677jK/9vTTT2P69OmoXr06fvjhBwwbNgwZGRnmL+AHH3wQDz74IHJycjBs2DAMGjQIbdu2df9x2yMxT+y5fv06iouLYTAYkJOTg+7du+O2225DTEwM5s2bh9TUVIwePRp//vkn7rjjDqxYsQLZ2dkoLCzEpk2b8Pjjj1tsT6/X47XXXsOZM2cQExODt956C3FxcVX2u3//fpSWliI8PNypQydyh9OnT+P06dPeToZLczWpBYMT8n2VpoYOBFD65pvA6NFAfDyuXLlifi0mJgabN282P4+KikK7du0AmH7df/LJJ+bX7r33Xtxzzz0ALKszevTogU6dOmHfvn0WdwfKtxEfH4+vvvrKu8GJjDyxp06dOnjkkUcAmI6tR48eyMjIQPPmzbFx40Z8+OGHAIBbbrkFkZGRqF27NvR6PYYPH45z586Zt5OTk4MGDRpg/vz5WLJkCeLj47Fw4UL0798fv//+O2666Sbzun///Tcee+wxLF26VIGMIFLOu+++i9mzZ3s7GZg5cyZmzZrl7WS4hMEJ+b5KU0O3BmAAMDA7G5+ePGnRbuH06dMICwvD6dOnodfrodPpUL9+fWRmZuL8+fNo0KABAgICsGfPHmRlZWHGjBnIzMxEXl4emjRpAsBUZbFv3z48/fTTyMzMxJ9//mlus3HhwgVs2rQJHTt2RGZmpsV+L126ZLEMcOMvIBl5Ys8DDzyAzZs34x//+Af+/vtv/Pjjj3jhhRfQoEED1K5dG9999x369euH/Px8nDhxAuvXr0e3bt0AAFu2bEFKSgqOHTuGHTt2IDw8HBMmTMC4ceMAAEuWLMHXX3+NFStW4MUXXwQAXL16FUOHDsWLL76I7t27203b1atXcfXqVfPzgoICiZlD5JynnnoKQ4YMceq9xcXF6NmzJwBg9+7dqFmzptPp0PpdEwDsraMmbHXvJunpQgDmxzeAaA2ITpGR4l//+pdo1KiROHbsmBBCiFmzZok2bdqIhg0bms+FNx8zZ870ep688soronnz5iIoKEg0atRING/eXPzxxx8CgDh27JgYPHiwaNeunWjXrp3Q6/XmXWzbtk107txZdOzYUURFRYnXX3/d/BoA0aNHD3HLLbeINm3aiA8++EAEBgaaewOVe+yxx8SQIUOEEKbeVCNHjpScJzNnzrSap75WdrDc8A3+cB7lfH/rhBBC+ZDHfQoKChASEgKj0Yh69ep5OzmKKioqQp06dQAAhYWFqF27tpdT5EMqt69ISQHmzbO5uit1x0r/AnLbryCZeVKRq59DnU6HjRs3YujQoQCAU6dOoXnz5tizZ4/5zgoATJ06FTt37oTBYMDu3btx5513omPHjubX33vvPXTo0MHqPqzdOQkPD/e5soPlhm/wh/Mop9xgtQ75B5lTQ7sSFFSsEomNjVVvIaOx6bJ79uyJsrIyyesHBwcjODjYjSkiIndhcEL+g1NDV6WSPAkNDUVgYCDy8vIslufl5Vk0NiYi/8BxTojI64KCgtClSxds377dvKysrAzbt2+3qOZxBicNJdIe3jkhIo8oLCzE0aNHzc+PHTuGrKwsNGzYEBEREUhOTkZiYiJuv/12xMXFYeHChSgqKjL33nFWUlISkpKSzPXdRKR+mglO9Ho99Ho9SktLvZ0UInLCvn37LCYGTE5OBgAkJiZi1apVGDFiBM6dO4cZM2bgzJkziI2NxebNm81dtInIf7C3jor4Q2ttf+AP51GLn0MtplkKf7je/IE/nEc5n0G2OSEiIiJVYXBCREREqsLghIh8GnvrEGkPgxMi8mlJSUk4fPgwMjIyvJ0UIpKIwQkRERGpCoMTIiIiUhUGJ0RERKQqDE6IyKexQSyR9jA4IQIAgwF47z3TX/IpbBBLpD0MTohSUoA77gAee8z0NyXF2ykiIvJrDE7IvxkMwIIFlssWLOAdFCIiL2Jw4gWsQVCR7Gx5y30NL0YiUiEGJx7GGgSViYyUt9yX8GIkIpVicOJBrEFQofh4YOpUy2UpKablvowXI/kL3h3UJAYnHuTvNQiqNX8+kJ4OrFlj+jtvnrdT5H5+dDGyK7Ef491BzWJw4kH+XIOgevHxwKOP+v4dk3J+dDGyK7Gf4t1BTWNw4kH+WoNAHiLn9jUvRvJ1fnR30BdV83YC/M38+cDw4abPR2QkvwtIISkplr8Sp041XWz28GIkX6bU3UGDgZ8RL/D4nZPc3Fz06dMH0dHR6NixIz7++GNPJ8Hr/K0GgdzMldvXvBjJVylxd5BtVrzG43dOqlWrhoULFyI2NhZnzpxBly5dMHDgQNSuXdvTSSHyDfZuXzPoIH/myt1BW0H/8OH8XHmAx4OTsLAwhIWFAQCaNm2K0NBQ/P333wxOyLN86VatHzVuJZItPt65zziDfq+SXa2za9cuDB48GM2aNYNOp8Nnn31WZR29Xo9WrVqhRo0aiI+Px48//mh1W/v370dpaSnCw8NlJ1xJ7AbvZ3ztVi0btxIpj0G/V8kOToqKihATEwO9Xm/19fXr1yM5ORkzZ85EZmYmYmJi0L9/f5w9e9Zivb///huPPfYY/vvf/zqXcoX42vcUOeCr3Qv9cawWiTjOCTmFQb93CRcAEBs3brRYFhcXJ5KSkszPS0tLRbNmzURqaqp52ZUrV0SvXr3EmjVrHO7jypUrwmg0mh+5ubkCgDAaja4kXQghRHq6EEDVR3q6y5t2SmFhoQAgAIjCwkLvJMLXrVlj/aRLuBal8ofzaDQaFfsceooW0yyFP1xvXpWebiof3PzF4A/nUc5nUNHeOiUlJdi/fz8SEhLMywICApCQkIC9e/eWB0MYO3Ys+vXrh0cffdThNlNTUxESEmJ+KFkFxG7wfoi3aolIDvZo8wpFg5P8/HyUlpaiSZMmFsubNGmCM2fOAAB++OEHrF+/Hp999hliY2MRGxuLQ4cO2dzmtGnTYDQazY/c3FzF0svvKT/EW7VERKrn8d46PXv2RFlZmeT1g4ODERwc7Ja0lH9PVWyCwO8pP8DBx4iIVE3R4CQ0NBSBgYHIy8uzWJ6Xl4emTZu6tG29Xg+9Xo/S0lKXtlOZnO8pX+p9aos/HCMA57sXKsVvMpqISD5Fq3WCgoLQpUsXbN++3bysrKwM27dvR7du3Vzatjsn75JSpegPvXr84RgteKsPud9lNNnDoQyIqpIdnBQWFiIrKwtZWVkAgGPHjiErKwsnTpwAACQnJ2Pp0qVYvXo1fv31VzzzzDMoKirCuHHjFE24J6ml96k7CzG1HKPHeCtA8LuMJnsYpxJZJzs42bdvHzp16oROnToBMAUjnTp1wowZMwAAI0aMwL///W/MmDEDsbGxyMrKwubNm6s0ktUSNfTqcXchpoZj9BhvBgh+ldFkD+NUIttktznp06cPhBB215k4cSImTpzodKLUxtu9ejwxxYO3j9GjvDkstV9lNNnD0dGJbPP4rMTO8uYoj97ufeqJH9vePkaP8maA4FcZrQ5qHSGWcSqRbTrh6DaIyhQUFCAkJARGoxH16tXz6L7d3cGiqKgIderUAWBq21M+GaLBYKrKqSw9Xfl0+E0nkpSUqn3IFRry3dZ5tKDxjPbm59BZakyzM5dh5UtH0vVGqucP51HOZ9Dj45xombd6n3pyPBZv97D1GG+PdeI3GU32yL0MKwczU6cC/2vu53Uaj7dJZRicaIS3v0t9EgMEUgGpl6GttmcDBrgnXXJYC5rmz/deekj7GJxoCL9LifyXrTZmR496Nh2VeaLBvqrxlpFbsEEsaQdHqyI/ZquhbOvWnk1HZX7dO54D1biNZoITd44Q6w38npWJhQD5OVsdvbz9e81vex1xoBq30kxw4kv4PSsTCwEiAKZ2HOnpwJo1pr8KdTBziU/3jrf3K9Kvbxm5H9uceJi9+tn27d2/b01WjXK0KiIzNbY988kG+45a+frCLSMVfynwzomHeSvY1vTdGl8oBIh8nJQJVDVDyt1ard8yUvmXAoMTD/PG96zma0W0XggQuYDt07xA6q9INdazSeHMl4KHL0TNBCe+0lvHG9+zPlE1qtVCgMgFKv9x67vk/IrU4i0juV8KXrgQNROc+FJvHU9/z/pMrYgWCwEiJ2n+jqeW+frdWjlfCl66EDUTnKiJEne3PPk96+ufMyJ7tHrX1SfueGqZL9+tlfOl4KULkb11ZNLqMM0+2ZqeSIKkpCQkJSWZJx3TCp+546llauwapRSpXwpeuhB550QGrd9mZa0IkXbwjie5nZQvBS9diLxzIgOH2yAiT1LqjqeKh7MgLfDCrXfNBCd6vR56vR6lpaVeSwNvs5Lm8VtKc1ytWdBqVTSpjIeruDRTraOG3jq8zUqaxn6pfkfrVdHkvzRz58QZJ04A+fnKbnPECCA62rTtiAigQwcgM1OZbRcX3/g/KwuoWVOZ7XpCaKgpP0il/H5ee//EqmjSKp8NTk6cAKKigMuXvZ0S5/Ts6e0UyFOrFvDrrwxQVIvfUn6JVdGkVT4bnOTnmwKT9983BSlaUFx8IyjZvVs7d05+/RV45BFTnjM4USl+S/ml8qroijfNWBVNWuCzwUm5qCigc2fP79eZdodFRTf+j40Fatd2S9LIl9m68Pgt5bc4xhFpkc8HJ97A1vHkUeUBybZtpqGLy1W+8Pgt5bd8eSwx8k0MThTGdofkUZUj4YqsXXj8liIiDdBMV2Kt4HwY5DHWIuHKeOGRHUrME0Y+SAUXhmaCE61M3sV2h+QxUgIPXnhkA4e9IatUcmFoJjhRwyBsUnCgNvIYR4EHLzyygYOzkVUqujDY5sQN2O6QPMJaD5xHHwXuvpsXHtnlN8PeqG26BrWlpzIVXRgMTtyE7Q7JIxgJkxO2brW+3KdqAdXWbVJt6bFGRe0SNFOtQ0Q2SJn2nOh/DAbT4JSV+dQlpKLqCVWmxxYVtUvgnRNSP3dMkuROnCRJVdQwo7ma2Lpzf/fdnk2HW6moesK8X1vL1RYRquRuLIMTUjdOkuRZPjhJUlJSEpKSklBQUICQkBBvJ8frVHTn3n3UdpBqS48jKmiXwOCE1M1TkyS9+SawevWN54mJwOTJzm1LTZMkHTpkOYW2PZwkyS/4xUwGajtItaVHAxickDa4c5Ikg8EyMAFMz595xrnCQy2TJGmhAR55hUru3LuX2g5SbelROQYn/6P2Hl7kRlqqD5aK8yiQAyq4c+9+ajtItaVHKi98QWqmt447R4hVyYB45C1aqw+WgvMoEJESvPQFqZngxF0jxGqlhxe5kYq6zynGFwMuDVPBVCWkJWoZCd2LX5CaCU7chT8wCYCpPjg9HVizxvR33jxvp8g1vhhwaRTvzJIk//rXjf/79lXHheLFL0i/b3PCH5hkptX6YFvYAM/r2PSHJDEYgIULLZep4ULx4hek39858fUfmErdTuZtaSukZoo3M4+jx3oV78ySJLYuiE2bvFvwevEL0u/vnAC++wNTqZ6k7JFqhdRM+de/LH8RMfP8Cu/MkiS2LojZs2/8762yw0tfkH5/56Scr/3AVKodExsMWyEnU6zdqvXrzPMvvn5nlhQSHw9MmWJ/HW+WHV74gvTb4MTXqikqH49St5N5W9oKR5lSsWGbnPeTT1JrW2tfKwM1b86cG/9Pm2Z9HT8qO/wyOPG11vPWjkep28m8LW2FvUyx1rBN6vvJZ6ntzqyvlYE+5557rC/3o7LD74ITX6umsHU8gDK3k3lb2gp7meLol43fZx55m6+VgT6pa1e/L3j9rkGsr41Ubu94lGrH5KsNhl1iK1Ns/bKZORMYMICZR17na2Wgz/LzgtfvghNfq6ZwdDxKDd3ha0OAKMJappQ3bKtYtZOSAsya5cGEkSdpbV4uXysDNUfOBePHBa/fVev4WjWFrx2PT6jYsC0tTZkWkGy9qEpabLvBMsOLtHjBeInf3TkBfO9uma8dj09RYqJKDjSjSloe/ZVlhhdo+YLxAr8MTgDfu1vma8dD/yO1QJNbt6C1uggVUrLthjdOB8sMD/O1xj5uvmg1U62j1+sRHR2Nrkr8EiXSCikDzci9Vcxby4pQqu0GT4ef8KXGPh64aDUTnCQlJeHw4cPIUMlU0v7SBMBfjtMl7swkWwVXScmNfcvpF2pr/VmzgEOHXEoqycduvX7EVxr7eOii1Uxwoib+8kvHX47TJe7OJGsFGgA88YRpX3KH8LW1fPZsYOxYp5Lor5QYPZkjMPsZtQ4XLIeHLloGJzLJCRq1fNeBv+gksJVJSt/dmz8fWLas6vIFC27cQanMlVvIvIMiiRJ36bV+p1/LZZzXqG24YEDeifTQReu3DWKdJbVNk9Y7WPha2y23sJVJR48qv6+gINvLp061vNjs3SouvxNTOaiq6MQJ59PpI06cAPLz7a9TvTqQmAisXn1j2dixpuWZmdL2o8Q2iotv/J+VBdSsKe19rnrzTct0JyYCkyfL305oKBARoVy6SCa5X1bWyhA3VE8xOJFJStDoCz3GtP6LziNsZUbr1p7bV2Sk6VeYnH6h5f1IN22ynJK9nIq/KYYNG4YdO3bgrrvuwoYNG9yyjxMngKgo4PJl+e9dtcr0cIUr2+jZ07V9u2L1astgRapatYBff3XTZSclylQLb0SZhw5Z/7KKjgY6dLD9vhEjTOucOGE6cR063PhfIQxOIH/APkdBo5buOtg6dg8Fx9pmK5Pc0aPM0QmR0i+08smOjzcViJULJ3uFkpc9++yzePzxx7HamW9BifLzTYHJ+++bghS1Ky6+EZTs3u2Z77SvvwZmzKi6/OWXgfvuk76dX38FHnnElOeKByeuRJne5s0oE3Cu/ZnCUabfByfOVL84GsBIK3cdHB07B2qSwFomFRV5bl9S2TrZFbdZVqb6RrF9+vTBjh07PLKvqCigc2eP7EoSWz8kKl5usbFA7druT8u1a9aDk2bN5OVZefOmQ4fckNf+HGUeOmR5V8PeetY+86tWyfuR4o4oU2iM0WgUAITRaLS73v79QgCmv7akp5vWqfxIT3c9nVOnWm4zJcXxewoLCwUAAUAUFha6ngg7lDx2KXntNLdu3AXp6UKsWWM1wySfRzvbUJTUky0jr6V+DivauXOnGDRokAgLCxMAxMaNG6uss3jxYtGyZUsRHBws4uLihMFgqLJOWlqaeOCBByTvV26a1XjJVS5Ppk698Zonyw17abKWNjnvl/o+ydR4Iu1Q7DzKzVhnvqwqk5jXcsoNv+6t484eUWrvMcYujC5QovuwJ/tpq+RkFxUVISYmBnq93urr69evR3JyMmbOnInMzEzExMSgf//+OHv2rFP7u3r1KgoKCiweWqTWnnP2OpE5Sptaj0nznMlYlX5Z+XVw4u7qFzX2GCunlaon1VGiVPV0yaySkz1gwADMmTMHw4YNs/r6G2+8gQkTJmDcuHGIjo7GkiVLUKtWLaxYscKp/aWmpiIkJMT8CA8PdyX5XqOS2NIqW53IHKVNzcekac5mrAq/rPw6OPGVAfucoeSxV6w39nlaHHlLAxd6SUkJ9u/fj4SEBPOygIAAJCQkYO/evU5tc9q0aTAajeZHbm6uUsn1KJXElrLS4Chtaj4mTfOhjPXr4ARQ7R0tj1Di2FNSbrSnGjvWD0aR1erIWyq/0PPz81FaWoomTZpYLG/SpAnOnDljfp6QkICHHnoI33zzDVq0aGE3cAkODka9evUsHlrkamzpzoHSnE2bBuJlbXJ3xnpw1D2/760DOO6F6csTuLoyM6kvjOcimxJ9rN3dT9te/3CNn5hvv/3W20nwCmc7anliMEhn0zZ/vmmojLFjTZ1DEhOVTZffclc3Sw+PLMrgxAGtj/TqTloaz8WthJD/Hh8pQJQSGhqKwMBA5OXlWSzPy8tD06ZNvZQqdZEbW3ryx4OzcW95b1UVD62jTUr/EPHCL1G/r9axhy3K7fOh6k3plLwolG6EpuELNigoCF26dMH27dvNy8rKyrB9+3Z069bNpW3r9XpER0ejqzsGx1MxNjolxXjhYmJwYgc/3PZ5tN740CF1zDCm5ovCVho2bVJF3hUWFiIrKwtZWVkAgGPHjiErKwsn/jeXT3JyMpYuXYrVq1fj119/xTPPPIOioiKMGzfOpf0mJSXh8OHDyFB6QkaV88sfD+QeXriYWK1jBz/cjnms3rjiKIberKpQ80VhKw0V58/xYt7t27cPffv2NT9PTk4GACQmJmLVqlUYMWIEzp07hxkzZuDMmTOIjY3F5s2bqzSS9WXlzYVKSkzddJ2p8avY5IhTUJAivDCfCYMTOzi/jDRurTe21j/Zm61u1XxRSJlx2It516dPHwgH7XMmTpyIiRMneihF6lK5uVA5OfHkv/4FLFxo+d70dN9t0E8e5OH5TBicOOCt+WUyMoA+fTyzL1X73y3/Klxpdetq9ytPXxRy0lsxbX/8YX3WYb9rsax+1poLlZMTT1YMTCq+99FHXU4ikUd7/HmlzcmwYcPQoEEDPPjgg97YvWyeGjzvX/+68X/fvn4wZogUtiaRcrYaRalh4z11UTiT3vK0DRhg/XU1VEF5kBYaxDo7oqoS2yZSI68EJ88++yzWrFnjjV2rlsFg/VePt9t/ep21uiJnq1G01pvF1fRypCsA2mgQ6+yIqkpsm0iNvBKc9OnTB3Xr1vXGrlVLzZ1AVGHVKtdHN9VaJiuRXpWPDEsm1uLIcnLiySlTnH8vkZrIbnOya9cuvPbaa9i/fz9Onz6NjRs3YujQoRbr6PV6vPbaazhz5gxiYmLw1ltvIS4uTqk0+yQ1dwKpyGuj5XboAHTu7No2tJLJ5ZRKrw+MDOsNX38NXLvmuayr2FzI2d46c+YAI0eqswGsz4y07TMHom6y75x4etpzfxEfr/5fPUo11/AarVVzaC29PuLNN01/Z8zw/HVe3lxo/HjnmzSpcIJZ7Zcd5XzmQDRAuACA2Lhxo8WyuLg4kZSUZH5eWloqmjVrJlJTUy3WS0tLEw888IDDfVy5ckUYjUbzIzc3VwAQRqPR7vv27xcCMP3VisLCQgFAABBpaYXeTo6F9HRTflZ+pKe7Oa/dsfH0dCHWrDH9dYOK57GwUIHz6Ob0msnIa6PRKOlzqCZS0mzvOlcrxa83hflM2eHmi0Pt59EuiXktp9xQtM2JO6Y9T01NRUhIiPkRHh6uVHJVTW0dC7TWXMMuNf60tEdr6VUZOb11fOo6VwmfyVOfORBtUDQ4cce059OmTYPRaDQ/cnNzFUuvB2d/9ui+lFIxzVprruETtHjRqJCc3jq8zpXnM3nqMweiDV7prfPtt9/i3LlzuHz5Mv766y+7E3sFBwejXr16Fg8leLLqUIvVlJXT/OmnbP7gUVq8aHxAfHzVKRh4nbvGZ5pO+cyBaIOiI8RqZdpzT87+7IWZpl1mK83p6d4ZLdfvaPGi8SGTJwOrVwMvvwzccw+zXAneGmlbcT5zIOqnaHBScdrz8u7F5dOeuzpfhl6vh16vR2lpqcvptFd1qPS15sl9KcVemtn0wQO0eNH4oPvuc733uhQGg2niaMA0qK8aT7ESafSZHu2ODkROV+OK67Zv71x67O1Pw92eZQcnhYWFOHr0qPl5+bTnDRs2REREBJKTk5GYmIjbb78dcXFxWLhwoWLTniclJaGgoAAhISEubcuTVYdK7EuJmUrl7scaTVWtevJDqfS+WLftNypP9jd7tncn3bZGC2lUjcqZZS+jKq9beSwJV/cnJy1qJLfHUFpamrm7U8VHYmKieZ233npLREREiKCgIBEXFyfSFeyHJ7UrkqOeTVOnWvYGS0lRLIlO78taV7LK7y1/TJ3q3jTGxUnPH9V1B6x8MEpnloN9KdIl0JMXaDl2JRZCeG4YAls9U+X2TnVnF1Sl0miL6soOV8jpamxl3cIK36eSzqO9/Xm6T7wbuhLLvnPiK9Oee7Lq0Nl9KTVTqTP7+fFHYNky996pcQs1NCiyNeGeHKzbVoySVcJKstcDVS01eFpIo2rIqY5VovuxM12bNXTSFG1z4k7uKGA8WQfqzL6kzFSqRPpt7ScoSINTrauhQVGFak+X+EwlvXcpWSWsJHu1dGqpwdNCGlVDTnWsEpnnTPWvhk6aV7oSO0MLM4sqzZ0zlUrZjoau4xvU0KCodWvl90U+x9Zkf2rqnaqFNKqGnK7G1tZ97jnl9ucD3Z41c+fEH8XHA488Arz/ftXXHn30xg93V6+38uu4Yg2Fxq7jG6wdjLu6GNnKOKWH99Vwi3uyr7z2Ts29dbSQRtWQUx1bed327YH//Ee5/Wm8apjBiYqlpFgGJvfeC4weDWzdaho49L33TMuVaISt8evY0vz5wKlTNzLvvfeAsDD3tFS3lnFFRcptX+st7skhLdTeaSGNqiEnsyqu62y5YW9/Gj5xDE5UKiOjalvLzZuBBx+seidFqfaeGr6OLRkM7sska9yVcRyMjVwk5aabo3WcuXGnmZt9njg4tWSGxtKtmTYncibv8gW22lTaanLDuacq8JUJunzlOMgrpMyA4GgdZ2ZR0MzMC544OLVkhhbTrVAvZ49R21gFSqo4XkFaWqHVburLlqlvSnfVjVXg5XnvFRt3QqVjFQjBcU7UxNr1JuXScbSOM5ef3Pd4rezwxMHJXN9t49W4Od1CCLeMc6KZOyf+pmtX642tx4/XfCNs9/OBluoAfOc4vMzf7roC0m66OVrH2WE05L7HKzxxcGrJDI2mm21OKnB2/gh3DS9vq5GqK41X1VL9qShrB6W1Fr62TozWjkOF1DrOiTtJ6VHvaB0lh9FQ3bAEnjg4tWSGVtPt8N6Kyrjr1qy1YeKljHqu5PDy7hyG2lpalRrV3avVOp4cql4i2edRLcfAah0hhG9U6wghbQYER+s4M4uCnPeoquxQ+uBkru/W8t+N6RZCuKVah8GJcH7+CHvvc/Rea7wxR4am58fwctsSW2SdRzUdA4MTIYTvBCdCmC6jNWscl2P21pGyDWff4/X2au48OJnru/vHqbvSLYRQx9w63uLO+TGcnYrAU8PLK8GTo7p7jC8clC8cA6mWlF7ujtZxpqe8ZoYl8MTBqSUzNJZuzTSIdefw9c5OReCp4eWVoJZqREX5wkH5wjEQESlMM8GJOzk7f4St90l5r6f5ZMcPXzgoXzgGP2QwmAYeNhiUWY9IFikXljMXn6P3ePCC1ky1jrs5O39Exc4USvfWUZpPdvzwhYPyhWPwI1JnFODMA+QWUi4sZy4+R+/x8AXN4KQCZ6vYXK2aK+9FGh4ubT1Xvr/UUv2pKKkHVTEDAemZuXy5aWjerl1NA81onU/2J/cMqTMKcOYBcgspF5YzF5+j93jhgma1jpdVHCW4b19p66l6SGi1qpyBUjMzPh544gng3XdNf93xQfTkyfXDC0nJQdikjk+lknGsyNcoMbqeM9v1wgXN4MSLrAWjUtdbsID12JI5ymhbmbl8OfDjj5bLfvzRtNydaXPXyfXTC0nJxvRS2y+znTO5hRKj6zmzXS9c0JoJTnxxCGqpQSd/hblISkZZW8fWl5mSPcY8eXJ5IblMavtltnMmt5ByYTlz8Tl6jxcuaM20OfHFIailBp38FeYiKRllbZ2uXU3VOdaWK8WTJ5cXkiKktl9mO2dyCykXljMXn6P3ePiC1sydE19kryuyo/X4K0wGRxltKzPHjwfi4qpuS8lGsZ48ubyQFBMfDzz6qLQBzqSsRySLlAvLmYvP0Xs8eEFr5s6Jr6oYjIaH224Uy19hLqqcgYC0zDQY3N9bx5MnlxcSEWkAgxMVKO8JW1QkbT1yUuUMlJqZ48e7vwuxJ08uLyQiUjmfD06+/hq4du1GV22pP5a9/cNy3TqgQwfH6XbHj/rK+1NDftglJYGuHoQSmeDOjFT9SSIikkHa9ITqIXVWw8REy0le4+KkzUrvzdnrK85KCRQ6THfl1+LiXE9D5eOXkm+qmvbcWgJdPaky3291dlF3XlhKbZuzEgshfGtWYrXz+qzEKqLl8+iOWYl9MjixNQu9o1npvT17fVpa1eDEVnqWLbP+2rJlzu/f2XzzWgEj5YS5elKdeH+VQsadF5aS22ZwIoTQ3Heapr/UGJzcoOXz6I7gRDPVOnq9Hnq9HqWlpQ7XlTN+SMU74N6evf7oUWnrZWfbH4LD2eodZ/PNa6ScMFdPqhIXhTsvLG9ftH5OTm2atXUNBvnzedmTkQH06eNamstfL58rrPKcYVqqHgcAHDoE/PKLa1W6mzYBf/0FtGihzIlyltxMdVd1tiem9FAqcPIU3jnhnROHCeadE945sULpOydyatOsrVt5maNt2DJlimW5ITcd9l6v/FC6etwjZYcrGWwrQ9xQ3+/wzonc6lt3VWdba0/Aah3n25zEx1s+T0mx/r7K58PWeu5grc2JvXRXvkbi411PQ+Xjl5JvqmpzYi2Brp5Ume+X1OZEyQtLqW0zOBFCSMsGOTGh1KDfmbjStO2q5YacdJSvKzedttIrJ2/cWnasWuVaBjvKEIV/tdoNTuT+CHHXj7J//tP68pdeUjw40Uy1jlyTJwOrVwMvvwzcc4/025FqGQZi6VLHvXXcMQSHteNXze1Za9w1WqKS71dqG97Ytg+QUyUslZzaNLmzA7irxtDRus7OYqC26nGzEydcS4ijDPHkAcnNVHdVZ+/da335L7843qZMPhuclLvvPqBzZ9P/Uod3UMMwEKNGAbVrO06PO4bgsDYciLfzwy4pCXT1IJTIBHdmpOpPkve4Y+oLOTMByJ0dQM76SqTD0dxuctOgmlkSIiKsL3d13hC521GC3Ex19STYWq9bNyAtrerydu2ATz+Vtm2JOHw9EZFMcmYCsLWutRkV5M4mEB8PTJniWjrsze1mbRuO9qWaWRI6dKi6TE5C7GWIpw9Ibqa6ehJsvX/uXOtTegwdKm27Mvj8nRMiIneQU5tma93hw13vrTNnDrBwoen/tDT7vXXkzO3mSm8dVdU0rloFBAS4VqWrht46cjPVXdXZ1toTZGY6d0x2MDghInKSnNo0a+sqXRsnZcJsR/t09XW567ldhw436vadoZoDgfy0uKs62wNTejA4cZFSQ70bDKbu+OUyMoD33wfOnwcGDZJ/Hcgdi6DyryOf5coJcsfPwIrbBVy7mGwNpGFt+5X3RUSkJtL6FamHmkZ5dGaod/vbsT3OiZyh6aV2b1eqC79mRnl0tt+/EsPXS9muKxeT1IE07J1wdiUWQmhuYFFNjyyqmbLDA7R8HjnOiVBPAePsgGX2t2N/EDYpA6xJ7d6uZBd+TRQwzvb7V2IQNjnbdeakODtAReVtMzgRQmjuO03TX2qaKDs8RMvn0R3BiWZ66+j1ekRHR6OrlEpVD5Az1LsS2wFsD1kvZXuVlyuZLk2QmjFKvc/Z9HgjLUptg4hIIZoJTpKSknD48GFkSPmG9gBvdJWXEpdJ7d6upi78HuFsv393DdrgzPvdOYCEz51wItIyzQQnamOtG7iUMQCkbMfWelIaxUrt3q6mLvwe4Wy/f3cN2uDoxMu5mOQMpGGNT55wItIy9tZxgVJDvZdv59AhYMIE07K0NOd760jt3m5tTAOf7rzhbL9/dw3aUHm7gPMXk72BNNhbh4g0hsGJi5Qa6j0+Hmjf/kZw0rWr46nP5aTL1fV8hisnyB0ZZe0CcnafUgbSsPU/EZGKsFqHiIiIVIXBCREREakKq3WIiCSy1wyo/LXffjO1FWvUCLjtNst1K09JImX7cpoerVtnGq3d0XpStrl8OfDVV6bj6NZNWps0g8FyriDA/txB5aNiHzrk2gjzsllLqL0McWVk6coZ4My2pLzH0XDftrYhdZhze6NZu4NSY7B4iq8OpCSEdgfh4UBKlrR6HjkIm4mtbLA3ULCjwXinTq066G/lUZ+lDPJrbaDgitdb+eCN9gYUlrLNyml1tL6UPHCUZ3JHpnZI6ol0dIBKjSxdftJtbMtmuSFl/46G+7a1DanDnNvLs6lTOUKsEAxO1IjBiSWtnkcGJybWssHeQMGuDNBbPuqznG1UHig4Lc36yNLWBhSWMuDxsmXy0yA3/c4O2CyLnBNpKyFKjyxtZ1tWyw0p+3e0L1snVMqJlrreqlWKBydsc0JEPk2J0aXtDc7ryuC65WNKytlG5XWPHpW2nr39VFzuzEjUctPvzsGOHe5cznpKjywtd10p+3e0L1snVOqAplLWO3FC2rZkYHBCRD5NidGl7Q3O68rguuXxkpxtVF63dWtp69nbT8XlzoxELTf97hzs2OHO5ayn9MjScteVsn9H+7J1QqUG61LWi4iQti0ZfD44OXQIeO89U1seewwGaeu5+j5r65cvq1h2Vvzf1j4qL1++HHj6aWD6dNv7WL7c8jWDAZg1y/SwdgwV9+FsHrmFK4mxd9DuOEilt2nrZDqzb1WdVPWyN1CwlFGeU1KAuLiq2yxvFCt1kF9rg/la++6wNeivlAGPx4+vmlZH25Y60rW9PPPIQMVST1Z5QpQcWbp8uZxtSdm/o+G+x4+3vg1ry62lz9p6ldfp0MH2685yWPGjMnLrjSs+pDbkcra9k+uz2lvWHdtrEFd5edOm1qsCrW2j/GGt0ZvURn4V1/N4mxNXWtJZOyhHjcZksqg7njJFkW3aTb+t7To6nsqvJyZKPpH+1uakXHq6EGvW2G7PsWaNEP/8pxBPPWX6W3ndZctMr9maYdza9u3tUwjL623p0kJJ7TYcbbM8rUOHCjF+vOl/R+uXb3fmTNOjvG1JxeeVlTdVWLXKcZplc3QiKyfU3gFKyTBb76ucATa2ZbetmpT9l69j62TZ2kbl5VLWq7wOG8RKP7jyi95eOyIhlG/v5Nqs9tYbtjnTPsmVh9RGfjKvS+dU3rgrLensHZStTHWihZ5FIaPQNh2mv/J2HeWTvW0xOHHf9ewGmm2ALdiYviItn0d3BCeaqdaR26jNVvscqQ25nG3v5ImGXu6emFlqIz+3N16Ts1NXE2wrU91xkM5uU85F6SifvHLyiIik0UxwIrdRm632OVIbcjnb3skTDb1c6HQgidRGfm5vvCZnp64m2FamuuMgnd2mnIvSUT555eQREUmjmeBELmvtc6Q25HK2vZPSs9rbahBnrX1SWJj0bVRMk7X1pTTy80jjNWtcaUln66DsNRpz9SCnTFFum44avjlqJOeood/Ysc6li4hIYT4/fP2qVUBAgP1Rdm3NNu+I3Pc5mtU+PBzo29e0LC3txqzE1t5jbVvlQ2NbGza7fN3KIxtbG13ZVnoB949YLImzJ6zie60dtCvbtWXOHGDkSOW2WTGNtoaptrautXUqv169uukDQ0TkbUq1h/EUX23UJoR2G0SxUZslrZ5HjhBrorVLTrPXm2DZUZGWz6NfN4glIiIi/8DghIiIiFSFwQkRERGpCoMTIiIiUhUGJ0RERKQqDE6IiIhIVRicEBERkaowOCEiIiJVYXBCREREqsLghIiIiFSFwQkRERGpCoMTIiIiUhUGJ0RERKQqDE6IiIhIVRicEBERkap4JTj56quv0LZtW7Rp0wbLli3zRhKISINYdhD5h2qe3uH169eRnJyMtLQ0hISEoEuXLhg2bBgaNWrk6aQQkYaw7CDyHx4PTn788Ue0a9cOzZs3BwAMGDAAW7duxahRo9y2T4MByM4GIiNNz8v/j4+X9r6SEiAoyPb7DQZg0yYgMxMIDASio4HbbruxTvl2tm8HDh0CSkuBkBDTo3NnYMAA4OefgT17bux77lzTPgcMqLofAAgPB3JzTf+Xr1P++l9/AWfPAkYj0KYNMGGCZTqsHXvlbQcFWR63rfcvXw588438cyLb3LnAwIHA+PG2D6TiQVTMuPKEZmQAXbsC7dtbrgdYv0CAqpli68KpeLEIYf318pPTokXV/VY8yZUvup9/vpH28eOr7tPRhW1tm9b2ZzCYlh06ZLowVcZbZUflS6r8UsrNBf74A6hZE5g4sepl5egzV3EfS5cC588DgwaZtlP5tJaUmPb32WfAxYtAnz5Ar17AV18BV6/e2Na6dUCHDlXLgxYtgGvXgMOHTeXPTTdZXobWPjaV0y71MrL1mrWP69atLpwcRw4dMv1dsuTG58baAUn9/JR/Fr/6CmjUyFSwln82c3OB4mKgWzdgyBDLkwZULY8qKr+gOna8sSwjw3SSHbF2kmyVgfbeV3lZeX5IOQZ3nUgh086dO8WgQYNEWFiYACA2btxYZZ3FixeLli1biuDgYBEXFycMBoP5tY8//lgkJSWZny9YsEC89tprkvdvNBoFAGE0Gu2ut3+/EIAQiYmmv9YeU6fafv/UqbbfV/H9jtaLi3O8nRuPQgHgf49CRfdT+fWKxy7lWK29v/Kydu0kn0bp2rWz3EnTptYPxNpBlL8m7yRIO/F2LpZC8zmEKCwsVO5iKj8Rjk6alJPraH/2PhxC+uewIl8oOypffs5+5spJOeWOH1XLDVcueWuXhq3t2buMrL1m6+OamCj5NErjSoEpt3CUk7GVVUiHRbkh4TMo6SRJueiklj3WtifzRMopN+BwjUq++eYbMX36dPHpp58KawXMhx9+KIKCgsSKFSvEL7/8IiZMmCDq168v8vLyhBDyC5grV64Io9FofuTm5ko6uPICxtEjPb3qe9PTlbse5T2sByfueqSnK3+sy5ZJv5YcWrZM+k5tvfbPf7ov82xcLBaFjF6v/L6lHJMSJ9fah+N/nAlOfK3scDVLlfvsebbccPVh6+Nq53KTR4mMdUfhWPkgK2VEleDEXqbISZvSF52dss/RiZRTbshuEDtgwADMmTMHw4YNs/r6G2+8gQkTJmDcuHGIjo7GkiVLUKtWLaxYsQIA0KxZM5w8edK8/smTJ9GsWTOb+0tNTUVISIj5ER4eLjfJdmVnS1vmi7KzlT/WjAwvbMzeenv3KpOWysozzlEG7t+v/L6lHJMSJ1fhi8PXyg5nVMxSfylnKrP1cVUsP5TYkDsKx/LtlpNSvtlKg5y0KX3ROSr7FMo3RXvrlJSUYP/+/UhISLixg4AAJCQkYO//CtS4uDj8/PPPOHnyJAoLC7Fp0yb079/f5janTZsGo9FofuSW138ppLx6zdEyXxQZqfyxdu3qhY3ZW69bN2XSUll5xjnKwC5dlN+3lGNS4uR68IOgxbLDGRWz1F/KmcpsfVwVyw8lNuSOwrF8u+WklG+20iAnbUpfdI7KPqXyTcpdMlsAy1uzJ0+eFADEnj17LNZ74YUXRFx5XbkQ4vPPPxdt2rQRt956q3j33Xdl7VPJNicpKbbfL6UKLiXF8Xrx8XLumFm/PavEfiq/XvHYpRyrtfdXruJs317WqZSmcpuTsDDrB2LtIMpfU7rNSeULR4k2J1JOcvmJcHTSpJxcR/uz9+EQzlXrVKTVsqPy5efsZ66cu9qcyCt3HF8atrZn7zKy9pqtj+vYsbJOpWOuFJhyC0c5GVuZvTYnDj6Dkk6SlItOatljbXsyT6ScckMnhBDOBjY6nQ4bN27E0KFDAQCnTp1C8+bNsWfPHnSr8Otu6tSp2LlzJwzlvQFcUFBQgJCQEBiNRtSrV8/mepmZph+t+/ebWqlro7dOEVasqAMAmDatEEFBtTXTW+fTT015rXgnj/ITOXy4ZnrrFAmBOhMmAAAKCwtRu3Zt7fTWWboUWLUKSEy0e1qkfg5t0VLZof7eOkXYtMlUbixdWogOHWprqrfOjBluKjtWrwbGjjUVgBrorVPUsSPqJCUBAArT0lBba711JJxIWeWGw/DFDsDy18/Vq1dFYGBglYZujz32mBgyZIgruxKLFy8WUVFRIjIyUlLkVf7rZ/9+l3brUYWFN34BFRYWejs5krk1rzV4IrV6HuXktdJ3TtxZdpSTe+dEK5ecZq83wbKjIi2fR6l57dYGsfYEBQWhS5cu2L59u3lZWVkZtm/fbvFryBlJSUk4fPgwMhRtcUlEauDOskOv1yM6OhpdFW0QRUTuJHsQtsLCQhw9etT8/NixY8jKykLDhg0RERGB5ORkJCYm4vbbb0dcXBwWLlyIoqIijBs3TtGEE5G2eKvsSEpKQlJSkvmWMhGpn+zgZN++fejbt6/5eXJyMgAgMTERq1atwogRI3Du3DnMmDEDZ86cQWxsLDZv3owmTZool2oi0hyWHUQklezgpE+fPhAO2tBOnDgREydOdDpRROR7WHYQkVRemZXYGaw3JiIi8g+aCU7YIJaInMEfNkTao5nghIjIGfxhQ6Q9DE6IiIhIVRicEBERkapoJjhhvTEREZF/0ExwwnpjIiIi/yB7nBNvKx8noaCgwO56hYU3/jpYVTWKiorM/xcUFKC0tNSLqZHOrXmtwROp1fMoJ6/LP3+Oxi1RA71eD71ej+vXrwPwvbJDs9cbWHZUpOXzKDWv5ZQbLs1K7A1//fUXwsPDvZ0MIgKQm5uLFi1aeDsZkrDsIFIHKeWG5oKTsrIynDp1CnXr1oVOp/N2chRXUFCA8PBw5ObmOjUVPamDr59HIQQuXbqEZs2aISBAG7XDvl52EKmdnHJDc8GJryufnMxoNPrkl5q/4HkkInKeNn7yEBERkd9gcEJERESqwuBEZYKDgzFz5kwEBwd7OynkAp5HIiLnsc0JERERqQrvnBAREZGqMDghIiIiVWFwQkRERKrC4ISIiIhUhcGJSuzYsQM6nQ4XL160u16rVq2wcOFCj6SJPGPWrFmIjY31djKIiFSDvXVUoqSkBH///TeaNGkCnU6HVatWYcqUKVWClXPnzqF27dqoVauWdxJKLtHpdNi4cSOGDh1qXlZYWIirV6+iUaNG3ksYEZGKaG5WYl8VFBSEpk2bOlyvcePGHkgNeVKdOnVQp04dbyeDiEg1WK0jQ58+fTBx4kRMnDgRISEhCA0NxUsvvWSe/vnChQt47LHH0KBBA9SqVQsDBgzAkSNHzO8/fvw4Bg8ejAYNGqB27dpo164dvvnmGwCW1To7duzAuHHjYDQaodPpoNPpMGvWLACW1TqjR4/GiBEjLNJ47do1hIaGYs2aNQBMk52lpqbi5ptvRs2aNRETE4MNGza4OafUp0+fPpg8eTKmTp2Khg0bomnTpuY8BYCLFy/iiSeeQOPGjVGvXj3069cPBw4csNjGnDlzcNNNN6Fu3bp44okn8OKLL1pUx2RkZODuu+9GaGgoQkJC0Lt3b2RmZppfb9WqFQBg2LBh0Ol05ucVq3W2bt2KGjVqVLlj9uyzz6Jfv37m57t370avXr1Qs2ZNhIeHY/LkyRZTrhMRaRmDE5lWr16NatWq4ccff8SiRYvwxhtvYNmyZQCAsWPHYt++ffjiiy+wd+9eCCEwcOBAXLt2DQCQlJSEq1evYteuXTh06BDmz59v9Rdz9+7dsXDhQtSrVw+nT5/G6dOn8fzzz1dZb8yYMfjyyy9RWFhoXrZlyxZcvnwZw4YNAwCkpqZizZo1WLJkCX755Rc899xzeOSRR7Bz5053ZI+qrV69GrVr14bBYMCCBQvw8ssvY9u2bQCAhx56CGfPnsWmTZuwf/9+dO7cGXfddRf+/vtvAMAHH3yAuXPnYv78+di/fz8iIiLwzjvvWGz/0qVLSExMxO7du5Geno42bdpg4MCBuHTpEgBT8AIAK1euxOnTp83PK7rrrrtQv359fPLJJ+ZlpaWlWL9+PcaMGQMA+OOPP3DvvffigQcewMGDB7F+/Xrs3r0bEydOVD7TiIi8QZBkvXv3FlFRUaKsrMy8LCUlRURFRYns7GwBQPzwww/m1/Lz80XNmjXFRx99JIQQokOHDmLWrFlWt52WliYAiAsXLgghhFi5cqUICQmpsl7Lli3Ff/7zHyGEENeuXROhoaFizZo15tdHjRolRowYIYQQ4sqVK6JWrVpiz549FtsYP368GDVqlOzj17LevXuLnj17Wizr2rWrSElJEd9//72oV6+euHLlisXrt956q3j33XeFEELEx8eLpKQki9d79OghYmJibO6ztLRU1K1bV3z55ZfmZQDExo0bLdabOXOmxXaeffZZ0a9fP/PzLVu2iODgYPO1MX78ePHkk09abOP7778XAQEBori42GZ6iIi0gndOZLrjjjug0+nMz7t164YjR47g8OHDqFatGuLj482vNWrUCG3btsWvv/4KAJg8eTLmzJmDHj16YObMmTh48KBLaalWrRoefvhhfPDBBwCAoqIifP755+Zf2EePHsXly5dx9913m9s11KlTB2vWrMEff/zh0r61qGPHjhbPw8LCcPbsWRw4cACFhYVo1KiRRT4dO3bMnE+///474uLiLN5f+XleXh4mTJiANm3aICQkBPXq1UNhYSFOnDghK51jxozBjh07cOrUKQCmuzb33Xcf6tevDwA4cOAAVq1aZZHW/v37o6ysDMeOHZO1LyIiNWKDWA964okn0L9/f3z99dfYunUrUlNT8frrr2PSpElOb3PMmDHo3bs3zp49i23btqFmzZq49957AcBc3fP111+jefPmFu/zxwnpqlevbvFcp9OhrKwMhYWFCAsLw44dO6q8pzwgkCIxMRHnz5/HokWL0LJlSwQHB6Nbt24oKSmRlc6uXbvi1ltvxYcffohnnnkGGzduxKpVq8yvFxYW4qmnnsLkyZOrvDciIkLWvoiI1IjBiUwGg8HieXnbgujoaFy/fh0GgwHdu3cHAJw/fx6///47oqOjzeuHh4fj6aefxtNPP41p06Zh6dKlVoOToKAglJaWOkxP9+7dER4ejvXr12PTpk146KGHzF/C0dHRCA4OxokTJ9C7d29XDtunde7cGWfOnEG1atXMjVQra9u2LTIyMvDYY4+Zl1VuM/LDDz/g7bffxsCBAwEAubm5yM/Pt1inevXqks7rmDFj8MEHH6BFixYICAjAfffdZ5Hew4cPo3Xr1lIPkYhIU1itI9OJEyeQnJyM33//HevWrcNbb72FZ599Fm3atMH999+PCRMmYPfu3Thw4AAeeeQRNG/eHPfffz8AYMqUKdiyZQuOHTuGzMxMpKWlISoqyup+WrVqhcLCQmzfvh35+fm4fPmyzTSNHj0aS5YswbZt28xVOgBQt25dPP/883juueewevVq/PHHH8jMzMRbb72F1atXK5sxGpaQkIBu3bph6NCh2Lp1K3JycrBnzx5Mnz4d+/btAwBMmjQJy5cvx+rVq3HkyBHMmTMHBw8etKjia9OmDd577z38+uuvMBgMGDNmDGrWrGmxr1atWmH79u04c+YMLly4YDNNY8aMQWZmJubOnYsHH3zQ4k5XSkoK9uzZg4kTJyIrKwtHjhzB559/zgaxROQzGJzI9Nhjj6G4uBhxcXFISkrCs88+iyeffBKAqRdGly5dMGjQIHTr1g1CCHzzzTfmOxmlpaVISkpCVFQU7r33XkRGRuLtt9+2up/u3bvj6aefxogRI9C4cWMsWLDAZprGjBmDw4cPo3nz5ujRo4fFa6+88gpeeuklpKammvf79ddf4+abb1YoR7RPp9Phm2++wZ133olx48YhMjISI0eOxPHjx9GkSRMApjyeNm0ann/+eXTu3BnHjh3D2LFjUaNGDfN2li9fjgsXLqBz58549NFHMXnyZNx0000W+3r99dexbds2hIeHo1OnTjbT1Lp1a8TFxeHgwYMWASdgajuzc+dOZGdno1evXujUqRNmzJiBZs2aKZgrRETewxFiZejTpw9iY2M5fDwBAO6++240bdoU7733nreTQkTkU9jmhEiCy5cvY8mSJejfvz8CAwOxbt06fPvtt+ZxUoiISDkMTogkKK/6mTt3Lq5cuYK2bdvik08+QUJCgreTRkTkc1itQ0RERKrCBrFERESkKgxOiIiISFUYnBAREZGqMDghIiIiVWFwQkRERKrC4ISIiIhUhcEJERERqQqDEyIiIlKV/welGhk8QMdHKgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Want: first seperate data by hostClass\n",
    "# Then seperate by zoonotic potential\n",
    "# Anthropod,Bacteria,Fish-like,Fungi,Mammal,Other,Parasite,Plant,Virus\n",
    "\n",
    "def make_panel(threshold):\n",
    "    zoonotic_count_matrix = pd.read_csv(neo4j_data_path + f\"{threshold}/\" + f\"count_matrix_positive.csv\", index_col=0)\n",
    "    not_zoonotic_count_matrix = pd.read_csv(neo4j_data_path + f\"{threshold}/\" + f\"count_matrix_negative.csv\", index_col=0)\n",
    "    total_count_matrix = pd.read_csv(neo4j_data_path + f\"{threshold}/\" + f\"count_matrix_total.csv\", index_col=0)\n",
    "    random_count_matrix = total_count_matrix.sample(n=100, random_state=1)\n",
    "\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "    plt.tick_params(\n",
    "        axis='x',          \n",
    "        which='both',      \n",
    "        bottom=False,      \n",
    "        top=False,         \n",
    "        labelbottom=False\n",
    "    )\n",
    "    axes = [ax1, ax2]\n",
    "    host_classes = [\n",
    "        'Mammal',\n",
    "        'Bacteria',\n",
    "    ]\n",
    "    for i in range(len(host_classes)):\n",
    "        host_class = host_classes[i]\n",
    "        ax = axes[i]\n",
    "        postive = zoonotic_count_matrix[host_class]\n",
    "        negative = not_zoonotic_count_matrix[host_class]\n",
    "        ys = [\n",
    "            postive,\n",
    "            negative,\n",
    "        ]\n",
    "        labels = [\n",
    "            zoonotic_count_matrix[host_class].index.tolist(),\n",
    "            not_zoonotic_count_matrix[host_class].index.tolist(),\n",
    "        ]\n",
    "        make_scatter_bar_plot(ys, labels, ax, host_class)\n",
    "    \n",
    "    if save_path is not None:\n",
    "        plt.savefig(save_path, bbox_inches=\"tight\")\n",
    "\n",
    "make_panel()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ZSEA Result investigation\n",
    "- original threshold attempted to filter out metagenomics using kmers/total spots\n",
    "- using kmers/total kmers percentIdentity instead increases majority taxOrder from primates to Bacteroidales (bacteria found in human gut) and Eubacteriales (?)\n",
    "- likely increases number of metagenomic samples (verify using host labels?)\n",
    "- u207044 is a majority class, 8973 mammallian, Influenza A virus\n",
    "\n",
    "\n",
    "\n",
    "The STAT query used above excluded more metagenome runs when the ratio of  (matched kmers / total spots) < 0.2, which artem suggested during previous work on TGaV to seperate 'Mixed' samples.\n",
    "\n",
    "In this context it seems okay to include them as long as (matched kmers / total kmers) > threshold . Doing so increases number "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Community detection and node centrality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_gds_projection(neo4j_data):\n",
    "    graph_name = 'zoonotic'\n",
    "    nodes = pd.concat([\n",
    "        neo4j_data['sotu_nodes'][['nodeId', 'labels', 'taxId']],\n",
    "        neo4j_data['taxon_order_nodes'][['nodeId', 'labels']],\n",
    "    ])\n",
    "    relationships = pd.concat([\n",
    "        neo4j_data['sotu_msa_edges'][['sourceNodeId', 'targetNodeId', 'relationshipType', 'weight']],\n",
    "        neo4j_data['has_host_order_stat_edges'][['sourceNodeId', 'targetNodeId', 'relationshipType', 'weight']],\n",
    "    ])\n",
    "\n",
    "    if gds.graph.exists(graph_name)['exists']:\n",
    "        gds.graph.drop(gds.graph.get(graph_name))\n",
    "\n",
    "    G = gds.alpha.graph.construct(\n",
    "        graph_name=graph_name,\n",
    "        nodes=nodes,\n",
    "        relationships=relationships,\n",
    "        concurrency=4,\n",
    "        undirected_relationship_types=['HAS_HOST_STAT'],\n",
    "    )\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_community_analysis(G):\n",
    "  communities = gds.labelPropagation.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    maxIterations=30,\n",
    "    # seedProperty='taxId',\n",
    "  )\n",
    "  unique_communities = communities.communityId.unique()\n",
    "  community_counter = collections.Counter(communities.communityId)\n",
    "  print('LPA Unique communities:', len(unique_communities))\n",
    "  print('LPA Most common communities:', community_counter.most_common(10))\n",
    "\n",
    "  wcc = gds.wcc.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    threshold=0.5,\n",
    "    # seedProperty='taxId',\n",
    "  )\n",
    "  unique_communities = wcc.componentId.unique()\n",
    "  community_counter = collections.Counter(wcc.componentId)\n",
    "  print('WCC Unique communities:', len(unique_communities))\n",
    "  print('WCC Most common communities:', community_counter.most_common(10))\n",
    "\n",
    "  page_ranks_sotu = gds.pageRank.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    maxIterations=100,\n",
    "  )\n",
    "  page_ranks_sotu = page_ranks_sotu.rename(columns={'score': 'pageRankSotu'})\n",
    "  page_ranks_sotu['pageRankRound'] = page_ranks_sotu['pageRankSotu'].round(0)\n",
    "\n",
    "  louvain = gds.louvain.stream(\n",
    "    G,\n",
    "    nodeLabels=['SOTU'],\n",
    "    relationshipWeightProperty='weight',\n",
    "    # seedProperty='taxId',\n",
    "  )\n",
    "  unique_communities = louvain.communityId.unique()\n",
    "  community_counter = collections.Counter(louvain.communityId)\n",
    "  print('Louvain Unique communities:', len(unique_communities))\n",
    "  print('Louvain Most common communities:', community_counter.most_common(10))\n",
    "  louvain = louvain.rename(columns={'communityId': 'louvainCommunityId'})\n",
    "\n",
    "  return communities, wcc, page_ranks_sotu, louvain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LPA Unique communities: 438\n",
      "LPA Most common communities: [(8170199, 45), (7858962, 39), (7743098, 27), (7935516, 20), (7999668, 15), (7971675, 13), (7714332, 12), (8100725, 12), (7994679, 12), (8017721, 10)]\n",
      "WCC Unique communities: 397\n",
      "WCC Most common communities: [(431, 86), (2, 42), (0, 40), (5, 23), (34, 17), (41, 15), (1, 14), (106, 14), (649, 12), (674, 11)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Louvain: 100%|| 100.0/100 [00:00<00:00, 231.77%/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Louvain Unique communities: 387\n",
      "Louvain Most common communities: [(435, 51), (208, 43), (163, 40), (604, 35), (5, 23), (126, 17), (188, 16), (80, 16), (193, 15), (24, 14)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'modularity' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[75], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m G \u001b[39m=\u001b[39m construct_gds_projection(neo4j_data)\n\u001b[0;32m----> 2\u001b[0m communities, wcc, page_ranks_sotu, louvain \u001b[39m=\u001b[39m run_community_analysis(G)\n\u001b[1;32m      3\u001b[0m G\u001b[39m.\u001b[39mdrop()\n",
      "Cell \u001b[0;32mIn[74], line 47\u001b[0m, in \u001b[0;36mrun_community_analysis\u001b[0;34m(G)\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mLouvain Most common communities:\u001b[39m\u001b[39m'\u001b[39m, community_counter\u001b[39m.\u001b[39mmost_common(\u001b[39m10\u001b[39m))\n\u001b[1;32m     45\u001b[0m louvain \u001b[39m=\u001b[39m louvain\u001b[39m.\u001b[39mrename(columns\u001b[39m=\u001b[39m{\u001b[39m'\u001b[39m\u001b[39mcommunityId\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mlouvainCommunityId\u001b[39m\u001b[39m'\u001b[39m})\n\u001b[0;32m---> 47\u001b[0m \u001b[39mreturn\u001b[39;00m communities, wcc, page_ranks_sotu, louvain, modularity\n",
      "\u001b[0;31mNameError\u001b[0m: name 'modularity' is not defined"
     ]
    }
   ],
   "source": [
    "G = construct_gds_projection(neo4j_data)\n",
    "communities, wcc, page_ranks_sotu, louvain = run_community_analysis(G)\n",
    "G.drop()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustering Notes:\n",
    "\n",
    "- General takeaways:\n",
    "    - Clustering SOTUs using weight mostly correlates to taxOrder, with some subcommunities within communities. Reovirales doesn't cluster well\n",
    "    - clustering with SOTU + Hosts creates on very large community and large number of small communities\n",
    "- \n",
    "\n",
    "- SOTU only, default community detection\n",
    "    - 742 nodes, 2714 edges\n",
    "    - LPA Unique communities: 438\n",
    "    - LPA Most common communities: [(8170199, 45), (7858962, 39), (7743098, 27), (7935516, 20), (7999668, 15), (7971675, 13), (7714332, 12), (8100725, 12), (7994679, 12), (8017721, 10)]\n",
    "    - WCC Unique communities: 386\n",
    "    - WCC Most common communities: [(431, 86), (2, 43), (0, 40), (5, 23), (34, 17), (48, 16), (79, 16), (41, 15), (1, 14), (106, 14)]\n",
    "    - Lieden Unique communities: 80\n",
    "    - Lieden Most common communities: [(145856, 215), (1239574, 49), (11303, 48), (291484, 48), (12637, 46), (11036, 25), (1679172, 24), (28876, 24), (147712, 22), (2593991, 21)]\n",
    "    - 356 stand alone nodes, 386 remain after hiding stand alone nodes\n",
    "    - standalone: zoonotic: 245/477, non-zoonotic: 141/265\n",
    "    - communities are mostly homogenous but contain some other colors (?)\n",
    "    - non-zoonotic are mixed in med-small communities, have med-small node degree\n",
    "    - zoonotic make up entire largest connected community\n",
    "    - \n",
    "\n",
    "- SOTU only, seeded community detection\n",
    "    - 739 nodes, 2570 edges\n",
    "    - LP Unique communities: 110\n",
    "    - LP Most common communities: [(145856, 215), (291484, 48), (11277, 47), (1239567, 37), (64300, 27), (28875, 20), (28344, 17), (147711, 17), (40058, 15), (11033, 15)]\n",
    "    - WCC Unique communities: 98\n",
    "    - WCC Most common communities: [(145856, 215), (291484, 45), (11072, 43), (11272, 41), (28875, 24), (40050, 23), (1239567, 22), (147711, 19), (1239566, 18), (11020, 16)]\n",
    "    - Lieden Unique communities: 80\n",
    "    - Lieden Most common communities: [(145856, 215), (1239574, 49), (11303, 48), (291484, 48), (12637, 46), (11036, 25), (1679172, 24), (28876, 24), (147712, 22), (2593991, 21)]\n",
    "    - 356 stand alone nodes, 383 remain after hiding stand alone nodes\n",
    "    - standalone: zoonotic: 242/477, non-zoonotic: 141/265\n",
    "    - communities are more homogenous than the default\n",
    "    - taxOrder and taxId corresponds nicely with clusters\n",
    "    - largest community is even larger\n",
    "    - overall zoonotic/non-zoonotic appears the same\n",
    "    - non-zoonotic exist in larger community now\n",
    "    - low-weight edges seem to involve long-range\n",
    "\n",
    "\n",
    "- Varying threshold:\n",
    "    - Varying threshold has little effect on seperating the largest communities and only seems to disconnect small communities\n",
    "    - No threshold, WCC Unique communities: 98\n",
    "    - 0.5 threshold, WCC Unique communities: 103\n",
    "    - 0.75 threshold, WCC Unique communities: 131\n",
    "    - 0.9 threshold, WCC Unique communities: 177\n",
    "\n",
    "\n",
    "- Including Hosts\n",
    "    - creates one very large community, increases number of overall communities\n",
    "    - LPA Unique communities: 269\n",
    "    - LPA Most common communities: [(8821550, 588), (7999668, 15), (8100725, 14), (7714332, 12), (8132309, 10), (7707513, 5), (7860156, 5), (7858962, 4), (7721352, 4), (8085484, 3)]\n",
    "    - WCC Unique communities: 285\n",
    "    - WCC Most common communities: [(0, 639), (108, 5), (155, 2), (162, 2), (3, 1), (12, 1), (18, 1), (21, 1), (23, 1), (27, 1)]\n",
    "\n",
    "\n",
    "- Page Rank:\n",
    "    - Visually seems to be some overlap between pagerank on SOTUs only and zoonotic potential\n",
    "    - including taxons made overlap less clear since taxon nodes seem to have higher centrality overall"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Graphistry visualizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create node and relationship dataframes with full information\n",
    "def get_graphistry_df(\n",
    "        neo4j_data, communities, page_ranks_sotu, louvain\n",
    "    ):\n",
    "\n",
    "    neo4j_data['sotu_nodes']['displayLabel'] = neo4j_data['sotu_nodes']['taxOrder']\n",
    "    neo4j_data['taxon_order_nodes']['displayLabel'] = neo4j_data['taxon_order_nodes']['taxOrder']\n",
    "    neo4j_data['sotu_nodes']['type'] = 'SOTU'\n",
    "    neo4j_data['taxon_order_nodes']['type'] = 'Taxon'\n",
    "\n",
    "    nodes = pd.concat([\n",
    "        neo4j_data['sotu_nodes'],\n",
    "        neo4j_data['taxon_order_nodes'],\n",
    "    ])\n",
    "    nodes = nodes.merge(\n",
    "        communities,\n",
    "        left_on='nodeId',\n",
    "        right_on='nodeId',\n",
    "        how='left',\n",
    "    )\n",
    "    nodes = nodes.merge(\n",
    "        page_ranks_sotu,\n",
    "        left_on='nodeId',\n",
    "        right_on='nodeId',\n",
    "        how='left',\n",
    "    )\n",
    "    nodes = nodes.merge(\n",
    "        louvain,\n",
    "        left_on='nodeId',\n",
    "        right_on='nodeId',\n",
    "        how='left',\n",
    "    )\n",
    "\n",
    "    nodes = nodes[[\n",
    "        'appId', 'labels', 'type',\n",
    "        'taxId', 'isZoonotic', 'hostClass',\n",
    "        'taxRank', 'communityId',\n",
    "        'taxOrder', 'displayLabel',\n",
    "        'pageRankSotu', 'louvainCommunityId',\n",
    "    ]].astype(str)\n",
    "\n",
    "    # Encode communityId to use default color pallette\n",
    "    nodes['communityId'] = nodes['communityId'].astype('float64')\n",
    "    nodes['communityId'] = nodes['communityId'].fillna(-1)\n",
    "    nodes['communityId'] = nodes['communityId'].astype('int32')\n",
    "    labels = nodes['communityId'].unique()\n",
    "    mapping = {label: i for i, label in enumerate(labels)}\n",
    "    nodes['communityId'] =  nodes['communityId'].replace(mapping)\n",
    "    nodes['communityColorCodes'] = nodes['communityId'].mod(11)\n",
    "\n",
    "\n",
    "    # Add host counts for palmprints\n",
    "    stat_hosts_excl_mixed = neo4j_data['has_host_order_stat_edges'][neo4j_data['has_host_order_stat_edges']['targetAppId'] != 12908]\n",
    "    host_counts = stat_hosts_excl_mixed.groupby('sourceAppId').size().groupby(level=0).max()\n",
    "    # host_counts = neo4j_data['has_host_order_stat_edges'].groupby('sourceAppId').size().groupby(level=0).max()\n",
    "    host_counts = host_counts.reset_index()\n",
    "    host_counts = host_counts.rename({0: 'hostCount'}, axis='columns')\n",
    "    host_counts['hostCountNormalized'] = round((host_counts['hostCount'] / host_counts['hostCount'].max()) * 100, -1).astype('int32')\n",
    "    nodes = nodes.merge(\n",
    "        host_counts,\n",
    "        left_on='appId',\n",
    "        right_on='sourceAppId',\n",
    "        how='left',\n",
    "    )\n",
    "\n",
    "    relationships = pd.concat([\n",
    "        neo4j_data['sotu_msa_edges'],\n",
    "        neo4j_data['has_host_order_stat_edges'],\n",
    "    ])\n",
    "    \n",
    "    relationships['targetAppId'] = relationships['targetAppId'].astype(str)\n",
    "    relationships['sourceAppId'] = relationships['sourceAppId'].astype(str)\n",
    "\n",
    "    relationships['weight'] = relationships['weight'].astype(float)\n",
    "    relationships['weightInt'] = round(relationships['weight'] * 100, -1).astype('int32')\n",
    "    relationships = relationships[[\n",
    "        'sourceAppId', 'targetAppId', 'relationshipType', \n",
    "        'weight', 'weightInt',\n",
    "    ]].astype(str)\n",
    "    relationships['weightInt'] = relationships['weightInt'].astype('int32')\n",
    "    return nodes, relationships\n",
    "\n",
    "\n",
    "\n",
    "def load_or_create_graphistry_df(node_filename, relationship_filename, use_cache=False):\n",
    "    if use_cache and os.path.exists(graphistry_data_path + node_filename) \\\n",
    "            and os.path.exists(graphistry_data_path + relationship_filename):\n",
    "        nodes = pd.read_csv(graphistry_data_path + node_filename)\n",
    "        edges = pd.read_csv(graphistry_data_path + relationship_filename)\n",
    "    else:\n",
    "        neo4j_data = get_neo4j_data()\n",
    "        G = construct_gds_projection(neo4j_data)\n",
    "        communities, wcc, page_ranks_sotu, louvain = run_community_analysis(G)\n",
    "        # binary_pvals, multi_pvals = run_pval_analysis(neo4j_data)\n",
    "        nodes, edges = get_graphistry_df(\n",
    "            neo4j_data, communities, page_ranks_sotu, louvain\n",
    "        )\n",
    "        nodes.to_csv(graphistry_data_path + node_filename, index=False)\n",
    "        edges.to_csv(graphistry_data_path + relationship_filename, index=False)\n",
    "    return nodes, edges\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes, relationships = load_or_create_graphistry_df('nodes.csv', 'edges.csv', True)\n",
    "# nodes, relationships = load_or_create_graphistry_df('nodes_v2.csv', 'edges_v2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_color_pallete = [\n",
    "    \"rgb(166, 206, 227)\",\n",
    "    \"rgb(31, 120, 180)\",\n",
    "    \"rgb(178, 223, 138)\",\n",
    "    \"rgb(51, 160, 44)\",\n",
    "    \"rgb(251, 154, 153)\",\n",
    "    \"rgb(227, 26, 28)\",\n",
    "    \"rgb(253, 191, 111)\",\n",
    "    \"rgb(255, 127, 0)\",\n",
    "    \"rgb(202, 178, 214)\",\n",
    "    \"rgb(106, 61, 154)\",\n",
    "    \"rgb(255, 255, 153)\", \n",
    "    \"#ffffff\",\n",
    "]\n",
    "\n",
    "categorical_colors = {}\n",
    "\n",
    "for communityId in nodes.communityColorCodes.unique():\n",
    "    categorical_colors[str(communityId)] = alt_color_pallete[communityId % len(alt_color_pallete)]\n",
    "\n",
    "categorical_colors[11] = \"#808080\"\n",
    "\n",
    "# for i, row in nodes.iterrows():\n",
    "#     if row['hostCount'] > 1:\n",
    "#         nodes.at[i,'communityColorCodes'] = 11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: <Response [400]>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/lukepereira/anaconda3/envs/rnalab/lib/python3.11/site-packages/graphistry/arrow_uploader.py\", line 362, in refresh\n",
      "    raise Exception(out.text)\n",
      "Exception: {\"non_field_errors\":[\"Token has expired.\"]}"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "            <iframe id=\"8601032f-13ed-4efd-b2e2-e5b9fc92165b\" src=\"https://hub.graphistry.com/graph/graph.html?dataset=d79e833cdce64c04a4ea81f03e7c4044&type=arrow&viztoken=0c081e7e-7ccb-4ea0-b1a3-5f4091d2e9b9&usertag=4ba963ab-pygraphistry-0.29.2&splashAfter=1698941089&info=True&play=2000&menu=True&showArrows=True&edgeOpacity=0.25&pointOpacity=1.0&linLog=True&compactLayout=True&strongGravity=True&dissuadeHubs=False&edgeInfluence=5&showLabelPropertiesOnHover=True&pointsOfInterestMax=10&backgroundColor=#000000&background=#000000\"\n",
       "                    allowfullscreen=\"true\" webkitallowfullscreen=\"true\" mozallowfullscreen=\"true\"\n",
       "                    oallowfullscreen=\"true\" msallowfullscreen=\"true\"\n",
       "                    style=\"width:100%; height:500px; border: 1px solid #DDD; overflow: hidden\"\n",
       "                    \n",
       "            >\n",
       "            </iframe>\n",
       "        \n",
       "            <script>\n",
       "                try {\n",
       "                  $(\"#8601032f-13ed-4efd-b2e2-e5b9fc92165b\").bind('mousewheel', function(e) { e.preventDefault(); });\n",
       "                } catch (e) { console.error('exn catching scroll', e); }\n",
       "            </script>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 316,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g = graphistry.bind()\n",
    "\n",
    "g = g.bind(\n",
    "    source='sourceAppId',\n",
    "    destination='targetAppId',\n",
    "    edge_weight='weight',\n",
    ").edges(relationships)\n",
    "\n",
    "g = g.bind(\n",
    "    node='appId',\n",
    "    point_label='displayLabel',\n",
    "    point_size=None,\n",
    "    # point_size='statCoverageInt',\n",
    ").nodes(nodes)\n",
    "\n",
    "\n",
    "params = {\n",
    "        'play': 2000,\n",
    "        'menu': True, \n",
    "        'info': True,\n",
    "        'showArrows': True,\n",
    "        # 'pointSize': 2.0, \n",
    "        # 'edgeCurvature': 0.5,\n",
    "        'edgeOpacity': 0.25, \n",
    "        'pointOpacity': 1.0,\n",
    "        # 'lockedX': False, 'lockedY': False, 'lockedR': False,\n",
    "        'linLog': True, \n",
    "        'compactLayout': True,\n",
    "        'strongGravity': True,\n",
    "        'dissuadeHubs': False,\n",
    "        'edgeInfluence': 5,\n",
    "        # 'precisionVsSpeed': 0, 'gravity': 1.0, 'scalingRatio': 1.0,\n",
    "        # 'showLabels': True, 'showLabelOnHover': True,\n",
    "        # 'showPointsOfInterest': True, 'showPointsOfInterestLabel': True, \n",
    "        'showLabelPropertiesOnHover': True,\n",
    "        'pointsOfInterestMax': 10,\n",
    "        'backgroundColor': '#000000',\n",
    "        'background': '#000000',\n",
    "      }\n",
    "\n",
    "g = g.settings(url_params=params)\n",
    "\n",
    "\n",
    "g = g.addStyle(\n",
    "    bg={\n",
    "        'color': 'white',\n",
    "})\n",
    "\n",
    "# g = g.encode_point_color(\n",
    "#     'communityIds',\n",
    "#     categorical_mapping=categorical_colors,\n",
    "#     default_mapping='grey', \n",
    "# )\n",
    "\n",
    "g = g.encode_point_color(\n",
    "    'communityColorCodes',\n",
    "    categorical_mapping=categorical_colors,\n",
    ")\n",
    "\n",
    "g.plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(urllib.parse.urlencode(params))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rnalab",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "448a416c49845ffbddc886562179757704d684dae284cc536cfa1eca10a7a7d0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
